<!DOCTYPE html><html lang="en"><head><link rel="stylesheet" data-href="https://fonts.googleapis.com/css2?family=Barlow:wght@400;600;700&amp;display=swap"/><style type="text/css">.fresnel-container{margin:0;padding:0;}
@media not all and (min-width:0px) and (max-width:299px){.fresnel-at-zero{display:none!important;}}
@media not all and (min-width:300px) and (max-width:494px){.fresnel-at-screen-300{display:none!important;}}
@media not all and (min-width:495px) and (max-width:874px){.fresnel-at-screen-495{display:none!important;}}
@media not all and (min-width:875px) and (max-width:1149px){.fresnel-at-screen-875{display:none!important;}}
@media not all and (min-width:1150px) and (max-width:1424px){.fresnel-at-screen-1150{display:none!important;}}
@media not all and (min-width:1425px){.fresnel-at-screen-1425{display:none!important;}}
@media not all and (max-width:299px){.fresnel-lessThan-screen-300{display:none!important;}}
@media not all and (max-width:494px){.fresnel-lessThan-screen-495{display:none!important;}}
@media not all and (max-width:874px){.fresnel-lessThan-screen-875{display:none!important;}}
@media not all and (max-width:1149px){.fresnel-lessThan-screen-1150{display:none!important;}}
@media not all and (max-width:1424px){.fresnel-lessThan-screen-1425{display:none!important;}}
@media not all and (min-width:300px){.fresnel-greaterThan-zero{display:none!important;}}
@media not all and (min-width:495px){.fresnel-greaterThan-screen-300{display:none!important;}}
@media not all and (min-width:875px){.fresnel-greaterThan-screen-495{display:none!important;}}
@media not all and (min-width:1150px){.fresnel-greaterThan-screen-875{display:none!important;}}
@media not all and (min-width:1425px){.fresnel-greaterThan-screen-1150{display:none!important;}}
@media not all and (min-width:0px){.fresnel-greaterThanOrEqual-zero{display:none!important;}}
@media not all and (min-width:300px){.fresnel-greaterThanOrEqual-screen-300{display:none!important;}}
@media not all and (min-width:495px){.fresnel-greaterThanOrEqual-screen-495{display:none!important;}}
@media not all and (min-width:875px){.fresnel-greaterThanOrEqual-screen-875{display:none!important;}}
@media not all and (min-width:1150px){.fresnel-greaterThanOrEqual-screen-1150{display:none!important;}}
@media not all and (min-width:1425px){.fresnel-greaterThanOrEqual-screen-1425{display:none!important;}}
@media not all and (min-width:0px) and (max-width:299px){.fresnel-between-zero-screen-300{display:none!important;}}
@media not all and (min-width:0px) and (max-width:494px){.fresnel-between-zero-screen-495{display:none!important;}}
@media not all and (min-width:0px) and (max-width:874px){.fresnel-between-zero-screen-875{display:none!important;}}
@media not all and (min-width:0px) and (max-width:1149px){.fresnel-between-zero-screen-1150{display:none!important;}}
@media not all and (min-width:0px) and (max-width:1424px){.fresnel-between-zero-screen-1425{display:none!important;}}
@media not all and (min-width:300px) and (max-width:494px){.fresnel-between-screen-300-screen-495{display:none!important;}}
@media not all and (min-width:300px) and (max-width:874px){.fresnel-between-screen-300-screen-875{display:none!important;}}
@media not all and (min-width:300px) and (max-width:1149px){.fresnel-between-screen-300-screen-1150{display:none!important;}}
@media not all and (min-width:300px) and (max-width:1424px){.fresnel-between-screen-300-screen-1425{display:none!important;}}
@media not all and (min-width:495px) and (max-width:874px){.fresnel-between-screen-495-screen-875{display:none!important;}}
@media not all and (min-width:495px) and (max-width:1149px){.fresnel-between-screen-495-screen-1150{display:none!important;}}
@media not all and (min-width:495px) and (max-width:1424px){.fresnel-between-screen-495-screen-1425{display:none!important;}}
@media not all and (min-width:875px) and (max-width:1149px){.fresnel-between-screen-875-screen-1150{display:none!important;}}
@media not all and (min-width:875px) and (max-width:1424px){.fresnel-between-screen-875-screen-1425{display:none!important;}}
@media not all and (min-width:1150px) and (max-width:1424px){.fresnel-between-screen-1150-screen-1425{display:none!important;}}</style><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin /><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><script async="" defer="" data-domain="napari.org" src="https://plausible.io/js/plausible.outbound-links.js"></script><script>window.plausible = window.plausible || function() { (window.plausible.q = window.plausible.q || []).push(arguments) }</script><title>using dask and napari to process &amp; view large datasets - napari</title><link href="/_static/pygments.css" rel="stylesheet"/><link href="/_static/togglebutton.css" rel="stylesheet" type="text/css"/><link href="/_static/copybutton.css" rel="stylesheet" type="text/css"/><link href="/_static/mystnb.css" rel="stylesheet" type="text/css"/><link href="/_static/sphinx-thebe.css" rel="stylesheet" type="text/css"/><link href="/_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" rel="stylesheet" type="text/css"/><link href="/_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" rel="stylesheet" type="text/css"/><meta name="twitter:card" content="summary_large_image"/><meta name="twitter:site" content="@napari_imaging"/><meta name="twitter:title" content="using dask and napari to process &amp; view large datasets"/><meta name="twitter:description" content="Often in microscopy, multidimensional data is acquired and written to disk in many small files,
each of which contain a subset of one or more dimensions from the complete dataset.
For example, in a..."/><meta name="twitter:creator" content="@napari_imaging"/><meta name="twitter:image" content="https://napari.org/_images/dask_repr.png"/><meta property="og:title" content="using dask and napari to process &amp; view large datasets"/><meta property="og:type" content="article"/><meta property="og:url" content="https://napari.org/tutorials/applications/dask.html"/><meta property="og:description" content="Often in microscopy, multidimensional data is acquired and written to disk in many small files,
each of which contain a subset of one or more dimensions from the complete dataset.
For example, in a..."/><meta property="og:site_name" content="napari.org"/><meta name="og:image" content="https://napari.org/_images/dask_repr.png"/><meta name="next-head-count" content="24"/><link rel="preload" href="/_next/static/css/785de95ba9df0d56e52c.css" as="style"/><link rel="stylesheet" href="/_next/static/css/785de95ba9df0d56e52c.css" data-n-g=""/><link rel="preload" href="/_next/static/css/7e93e5fc63e5e1e15f1f.css" as="style"/><link rel="stylesheet" href="/_next/static/css/7e93e5fc63e5e1e15f1f.css" data-n-p=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-a40ef1678bae11e696dba45124eadd70.js"></script><script src="/_static/documentation_options.js" id="/_static/documentation_options.js" defer="" data-nscript="beforeInteractive"></script><script src="/_static/jquery.js" id="/_static/jquery.js" defer="" data-nscript="beforeInteractive"></script><script src="/_static/underscore.js" id="/_static/underscore.js" defer="" data-nscript="beforeInteractive"></script><script src="/_next/static/chunks/webpack-613fd858cdb9cf2af3be.js" defer=""></script><script src="/_next/static/chunks/framework-2191d16384373197bc0a.js" defer=""></script><script src="/_next/static/chunks/main-1f2c591c5d3bfcfc95e6.js" defer=""></script><script src="/_next/static/chunks/pages/_app-0d478e281568d2da0774.js" defer=""></script><script src="/_next/static/chunks/29107295-2648cb5e919f7c78c7cc.js" defer=""></script><script src="/_next/static/chunks/822-f820657a4eb0d83be054.js" defer=""></script><script src="/_next/static/chunks/pages/%5B%5B...parts%5D%5D-52dfcf2f9b1bb0d33228.js" defer=""></script><script src="/_next/static/wooSd9gyc-BMXR2p6Uzhj/_buildManifest.js" defer=""></script><script src="/_next/static/wooSd9gyc-BMXR2p6Uzhj/_ssgManifest.js" defer=""></script><style data-href="https://fonts.googleapis.com/css2?family=Barlow:wght@400;600;700&display=swap">@font-face{font-family:'Barlow';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHpv4kjgoGqM7EPCA.woff) format('woff')}@font-face{font-family:'Barlow';font-style:normal;font-weight:600;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E30-8c4w.woff) format('woff')}@font-face{font-family:'Barlow';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E3t-4c4w.woff) format('woff')}@font-face{font-family:'Barlow';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHpv4kjgoGqM7E_A8s5ynghnQci.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Barlow';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHpv4kjgoGqM7E_Ass5ynghnQci.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Barlow';font-style:normal;font-weight:400;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHpv4kjgoGqM7E_DMs5ynghnQ.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}@font-face{font-family:'Barlow';font-style:normal;font-weight:600;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E30-8s6Fostz0rdom9.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Barlow';font-style:normal;font-weight:600;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E30-8s6Vostz0rdom9.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Barlow';font-style:normal;font-weight:600;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E30-8s51ostz0rdg.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}@font-face{font-family:'Barlow';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E3t-4s6Fostz0rdom9.woff2) format('woff2');unicode-range:U+0102-0103,U+0110-0111,U+0128-0129,U+0168-0169,U+01A0-01A1,U+01AF-01B0,U+1EA0-1EF9,U+20AB}@font-face{font-family:'Barlow';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E3t-4s6Vostz0rdom9.woff2) format('woff2');unicode-range:U+0100-024F,U+0259,U+1E00-1EFF,U+2020,U+20A0-20AB,U+20AD-20CF,U+2113,U+2C60-2C7F,U+A720-A7FF}@font-face{font-family:'Barlow';font-style:normal;font-weight:700;font-display:swap;src:url(https://fonts.gstatic.com/s/barlow/v5/7cHqv4kjgoGqM7E3t-4s51ostz0rdg.woff2) format('woff2');unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}</style></head><body><div id="__next"><div class="min-h-screen flex flex-col"><div class="fresnel-container fresnel-lessThan-screen-1150 "></div><header class="bg-napari-primary h-napari-app-bar justify-center items-center px-6 screen-495:px-12 grid grid-cols-3 screen-1150:grid-cols-4 screen-1425:grid-cols-napari-5 justify-center gap-6 screen-495:gap-12"><nav class="flex items-center"><a href="/" class="inline-flex items-center"><svg class="w-4 h-4 screen-875:w-6 screen-875:h-6 mr-1" width="30" height="30" viewBox="0 0 30 30" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M27.8387 0H2.16127C0.967636 0 0 0.967636 0 2.16127V27.8387C0 29.0324 0.967636 30 2.16127 30H27.8387C29.0324 30 30 29.0324 30 27.8387V2.16127C30 0.967636 29.0324 0 27.8387 0Z" fill="#26283D"></path><path d="M27.7779 1.03139H2.22356C1.56514 1.03139 1.03139 1.56514 1.03139 2.22357V27.7779C1.03139 28.4363 1.56514 28.9701 2.22356 28.9701H27.7779C28.4363 28.9701 28.9701 28.4363 28.9701 27.7779V2.22357C28.9701 1.56514 28.4363 1.03139 27.7779 1.03139Z" fill="url(#paint0_radial)"></path><path d="M17.4655 27.3361C17.099 27.3361 16.7542 27.3201 16.4124 27.2897C16.0096 27.2645 15.61 27.202 15.2187 27.1028C14.4296 26.8759 13.6918 26.4986 13.0459 25.9918C12.1394 25.2971 11.3826 24.4265 10.8209 23.4322C10.5949 23.0338 10.3863 22.4775 9.83293 20.9899C9.71125 20.661 9.61999 20.4104 9.54321 20.2091C9.36359 19.7209 9.3042 19.5558 9.15789 19.2385C8.96705 18.7853 8.73274 18.3516 8.45823 17.9435C8.12277 17.4768 7.73384 17.0509 7.29937 16.6745C7.13279 16.5297 7.05601 16.4616 6.91405 16.3515C6.73732 16.2067 6.49396 16.0183 6.02752 15.6098C5.2395 14.9174 4.94254 14.5958 4.66876 14.2453C4.25597 13.7417 3.92726 13.1748 3.69532 12.5664C3.49242 12.0272 3.37416 11.4599 3.34476 10.8846C3.32978 10.3829 3.38289 9.88145 3.50265 9.39401C3.69467 8.45513 4.01235 7.54642 4.44713 6.69241C4.7481 6.05734 5.13833 5.46858 5.60599 4.94398C6.16001 4.34736 6.8007 3.83754 7.50652 3.43167C7.8302 3.24114 8.16941 3.07832 8.52052 2.94495C9.04221 2.75721 9.59283 2.66258 10.1473 2.66537C10.6436 2.66652 11.1376 2.73375 11.6161 2.86528C12.5756 3.10049 13.433 3.64024 14.0599 4.40366C14.7204 5.22066 14.987 6.2477 15.2028 7.63978C15.2578 8.00193 15.2796 8.18879 15.2984 8.35248C15.31 8.46547 15.323 8.57846 15.352 8.77836C15.5273 9.97199 15.6142 10.5167 15.8141 10.9324C16.1038 11.5408 16.6572 11.8957 17.0208 12.1304C17.9131 12.7098 18.8344 12.788 19.4442 12.8416C19.6847 12.8633 19.9585 12.8749 20.2569 12.8749C20.703 12.8749 21.0739 12.8503 21.2115 12.8402C21.4853 12.8199 21.7141 12.7938 21.9358 12.7677C22.272 12.7223 22.6106 12.6972 22.9498 12.6924C23.3746 12.6921 23.7976 12.7491 24.2071 12.8619C25.0966 13.067 25.8989 13.547 26.5002 14.2337C27.1101 14.958 27.2738 15.6721 27.4795 16.5804C27.6193 17.187 27.707 17.8044 27.7417 18.4259C27.7764 19.1313 27.8242 20.0946 27.4925 21.2303C27.1727 22.2747 26.6228 23.234 25.8831 24.0377C25.1073 24.908 24.1734 25.6233 23.1309 26.1453C22.5393 26.4354 21.9207 26.6669 21.2839 26.8363C20.4006 27.0835 19.4935 27.2365 18.578 27.2926C18.1912 27.3216 17.8175 27.3361 17.4655 27.3361Z" fill="#26283D"></path><path d="M17.4655 26.2265C17.1323 26.2265 16.8194 26.2134 16.5109 26.1859C16.1688 26.1658 15.8293 26.1139 15.4969 26.0309C14.8468 25.8412 14.2394 25.5279 13.7079 25.1082C12.924 24.5046 12.2698 23.7494 11.7842 22.8875C11.6016 22.5645 11.3873 21.9908 10.8701 20.6031C10.7484 20.2786 10.6586 20.0323 10.5804 19.8295C10.4008 19.3385 10.3269 19.1342 10.1618 18.7764C9.94359 18.2639 9.6768 17.7735 9.36504 17.3119C8.98218 16.7792 8.53831 16.293 8.04249 15.8633C7.84694 15.6852 7.74264 15.6026 7.59778 15.4882C7.41961 15.3433 7.19942 15.1738 6.75471 14.7842C6.03042 14.1497 5.76968 13.8672 5.53791 13.5703C5.19569 13.1552 4.92274 12.6875 4.7296 12.1854C4.41134 11.3736 4.36059 10.4814 4.58474 9.63882C4.75558 8.79295 5.04066 7.97424 5.43216 7.20521C5.68904 6.6581 6.02321 6.15074 6.42443 5.69869C6.89941 5.18449 7.44931 4.74497 8.05553 4.39498C8.32534 4.23625 8.6078 4.1001 8.90005 3.98793C9.29948 3.84394 9.72125 3.77185 10.1458 3.77499C10.5448 3.77727 10.9417 3.83182 11.3264 3.93723C12.0597 4.10871 12.7166 4.51577 13.1965 5.09609C13.6311 5.6422 13.8846 6.37663 14.1062 7.80492C14.1584 8.14389 14.1772 8.31048 14.1946 8.47127C14.2076 8.5944 14.2221 8.71753 14.254 8.93481C14.4336 10.1531 14.5321 10.8252 14.8146 11.4104C15.2492 12.2984 15.9734 12.7735 16.4182 13.056C17.5394 13.7803 18.6692 13.8817 19.3443 13.9425C19.6166 13.9672 19.9237 13.9788 20.2554 13.9788C20.7393 13.9788 21.142 13.9527 21.2941 13.9411C21.5838 13.9194 21.8387 13.8904 22.0604 13.8643C22.3529 13.8246 22.6474 13.8019 22.9425 13.7962C23.2691 13.7958 23.5942 13.8401 23.9087 13.9281C24.2796 14.028 25.0502 14.238 25.647 14.9421C26.0816 15.4592 26.2062 16.001 26.393 16.8252C26.5186 17.3689 26.5976 17.9223 26.6292 18.4795C26.661 19.14 26.7016 19.9614 26.422 20.9174C26.1469 21.8088 25.6746 22.6267 25.0401 23.3105C24.3582 24.0719 23.5378 24.6968 22.6224 25.1516C22.0928 25.4117 21.5387 25.6184 20.9681 25.7687C20.1555 25.9947 19.3214 26.1345 18.4795 26.1859C18.1333 26.2192 17.7929 26.2265 17.4655 26.2265Z" fill="#A59678"></path><path d="M9.21149 4.83535C10.0328 4.53405 10.8513 4.74409 11.0946 4.80783C11.6453 4.93454 12.1384 5.24019 12.4969 5.67697C12.8199 6.07533 13.0212 6.69098 13.2211 7.95558C13.3124 8.5437 13.295 8.56978 13.366 9.07823C13.5529 10.3472 13.6644 11.1062 14.0077 11.816C14.548 12.9314 15.4563 13.521 15.9329 13.8267C17.2366 14.6726 18.5403 14.7885 19.2646 14.8537C19.9624 14.9053 20.663 14.9053 21.3607 14.8537C22.4718 14.7711 22.9121 14.6045 23.6712 14.8117C24.0058 14.9015 24.549 15.0579 24.9546 15.536C25.2443 15.8851 25.337 16.2603 25.5109 17.0324C25.625 17.5246 25.6967 18.0257 25.7253 18.5302C25.7571 19.169 25.7904 19.86 25.5543 20.6697C25.3157 21.4363 24.906 22.1387 24.3563 22.7238C23.7531 23.3969 23.027 23.9485 22.2168 24.3491C21.736 24.5834 21.2327 24.7683 20.7146 24.901C19.9602 25.1106 19.1857 25.2397 18.4042 25.2863C17.7992 25.3373 17.1911 25.3373 16.5862 25.2863C16.2975 25.2698 16.0109 25.2267 15.7301 25.1574C15.1926 24.9997 14.6912 24.7384 14.254 24.3882C13.5654 23.8599 12.9908 23.1976 12.5649 22.4413C12.4201 22.1806 12.1811 21.5476 11.7088 20.2801C11.2916 19.1603 11.2323 18.943 10.9845 18.3969C10.7451 17.8345 10.451 17.297 10.1067 16.7919C9.68854 16.2019 9.20233 15.6632 8.65814 15.1869C8.19314 14.7639 8.16417 14.8044 7.35442 14.0961C6.64751 13.4761 6.43168 13.2342 6.25205 13.0053C5.96697 12.6629 5.73947 12.2764 5.57846 11.8609C5.45202 11.5217 5.37825 11.1651 5.35973 10.8035C5.35202 10.4734 5.39004 10.1439 5.47272 9.82425C5.62675 9.05475 5.88556 8.31001 6.24191 7.61082C6.46342 7.13752 6.75173 6.69847 7.09802 6.3071C7.32105 6.06667 7.56328 5.84479 7.82231 5.64366C8.03967 5.47474 8.26863 5.3213 8.50748 5.18446C8.73275 5.05002 8.96813 4.93329 9.21149 4.83535V4.83535Z" fill="url(#paint1_linear)"></path><defs><radialGradient id="paint0_radial" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse" gradientTransform="translate(15 15) scale(28.521)"><stop stop-color="#746B7F"></stop><stop offset="0.12" stop-color="#615B6F"></stop><stop offset="0.26" stop-color="#514C60"></stop><stop offset="0.37" stop-color="#4B475B"></stop><stop offset="0.44" stop-color="#4F4B5F"></stop><stop offset="0.51" stop-color="#5A566C"></stop><stop offset="0.59" stop-color="#6D6881"></stop><stop offset="0.67" stop-color="#87829E"></stop><stop offset="0.74" stop-color="#A9A2C3"></stop><stop offset="0.78" stop-color="#BCB5D9"></stop></radialGradient><linearGradient id="paint1_linear" x1="5.54949" y1="7.79624" x2="23.9884" y2="23.7393" gradientUnits="userSpaceOnUse"><stop offset="0.05" stop-color="#448491"></stop><stop offset="0.15" stop-color="#4A7986"></stop><stop offset="0.33" stop-color="#506E7B"></stop><stop offset="0.52" stop-color="#526A77"></stop><stop offset="0.6" stop-color="#56707C"></stop><stop offset="0.7" stop-color="#607F8B"></stop><stop offset="0.82" stop-color="#7099A3"></stop><stop offset="0.94" stop-color="#87BDC4"></stop><stop offset="1" stop-color="#94D1D6"></stop></linearGradient></defs></svg><span class="font-bold">napari</span></a></nav><div class="flex items-center w-full justify-self-end col-span-2 screen-1150:col-span-3"><form data-testid="searchInputForm" class="flex flex-auto items-center border-b-2 border-black"><input aria-label="Search input for searching napari.org site." data-testid="searchInputTextInput" class="flex flex-auto border-none outline-none bg-transparent w-0" value=""/><button aria-label="Submit search query" type="button"><svg class="h-5 w-5" width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M8.32004 7.68001L0.746704 15.2533M10.6667 9.60001C13.0134 9.60001 14.9334 7.68001 14.9334 5.33335C14.9334 2.98668 13.0134 1.06668 10.6667 1.06668C8.32004 1.06668 6.40004 2.98668 6.40004 5.33335C6.40004 7.68001 8.32004 9.60001 10.6667 9.60001Z" stroke="black" stroke-width="2"></path></svg></button></form><div class="fresnel-container fresnel-lessThan-screen-1150 "><button class="ml-6 flex lg:hidden" type="button"><svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Icon for opening side menu.</title><path d="M16 1H0" stroke="black" stroke-width="1.98857"></path><path d="M16 7.99924H0" stroke="black" stroke-width="1.98857"></path><path d="M16 15H0" stroke="black" stroke-width="1.98857"></path></svg></button></div></div></header><main class="mt-6 flex-grow"><div class="grid grid-cols-3 screen-1150:grid-cols-napari-4 screen-1425:grid-cols-napari-5 justify-center gap-6 screen-495:gap-12 px-6 pt-4 screen-495:px-12 screen-495:pt-9"><div class="fresnel-container fresnel-greaterThanOrEqual-screen-1150 "><ul class="space-y-4"><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/community/index.html">Community</a></span></div></li><li class=""><div class="flex items-center text-base font-bold"><span class="truncate overflow-ellipsis"><a href="/tutorials/index.html">Tutorials</a></span></div></li><li class="MuiCollapse-root MuiCollapse-entered" style="min-height:0px"><div class="MuiCollapse-wrapper"><div class="MuiCollapse-wrapperInner"><ul><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-xs font-bold uppercase tracking-widest border-transparent"><span class="truncate overflow-ellipsis"><p>Fundamentals</p></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/installation.html">napari installation tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/getting_started.html">getting started with napari</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/viewer.html">napari viewer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/image.html">image layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/labels.html">labels layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/points.html">points layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/shapes.html">shapes layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/surface.html">surface layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/tracks.html">tracks layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/fundamentals/vectors.html">vectors layer tutorial</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-xs font-bold uppercase tracking-widest border-transparent"><span class="truncate overflow-ellipsis"><p>Applications</p></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/applications/annotate_segmentation.html">annotating segmentation with text and bounding boxes</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/applications/annotate_points.html">annotating videos with napari</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary font-semibold border-black"><span class="truncate overflow-ellipsis"><a href="/tutorials/applications/dask.html">using dask and napari to process &amp; view large datasets</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/applications/napari_imageJ.html">Napari + ImageJ How-to-Guide</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/applications/cell_tracking.html">single cell tracking with napari</a></span></div></li><li class="border-l border-black py-1"><div class="flex items-center border-l-4 py-1 pl-4 text-sm hover:border-napari-primary border-transparent"><span class="truncate overflow-ellipsis"><a href="/tutorials/further-resources/napari-workshops.html">napari workshops</a></span></div></li></ul></div></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/plugins/stable/index.html">Plugins</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/release/index.html">Release notes</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/api/stable/index.html">API reference</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/roadmaps/index.html">Roadmaps</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/guides/stable/index.html">Developer guides</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="/developers/index.html">Developer resources</a></span></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="https://github.com/napari/napari" target="_blank" rel="noreferrer">napari source code</a></span><svg class="ml-2 w-3 h-3" width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M10.6667 2H2V18H18V10" stroke="#000" stroke-width="2.5"></path><path d="M12.6667 2H18V7.33333" stroke="#000" stroke-width="2.5"></path><path d="M18 2L7.33333 12.6667" stroke="#000" stroke-width="2.5"></path></svg></div></li><li class=""><div class="flex items-center text-base"><span class="truncate overflow-ellipsis"><a href="https://napari-hub.org" target="_blank" rel="noreferrer">napari hub</a></span><svg class="ml-2 w-3 h-3" width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M10.6667 2H2V18H18V10" stroke="#000" stroke-width="2.5"></path><path d="M12.6667 2H18V7.33333" stroke="#000" stroke-width="2.5"></path><path d="M18 2L7.33333 12.6667" stroke="#000" stroke-width="2.5"></path></svg></div></li></ul></div><div class="col-span-3 mb-6 screen-495:mb-12 screen-1425:col-start-2 screen-1425:col-span-3 overflow-x-auto"><h1 class="text-5xl leading-tight font-bold mb-3 screen-875:mb-10">using dask and napari to process &amp; view large datasets</h1><div class="fresnel-container fresnel-lessThan-screen-1425 "><div class="fresnel-container fresnel-greaterThanOrEqual-screen-1425 sticky top-12"></div><div class="fresnel-container fresnel-lessThan-screen-1425 "><div class="transition-all mb-0"><button class="MuiButtonBase-root MuiButton-root MuiButton-text pl-0 transition-all mb-0" tabindex="0" type="button"><span class="MuiButton-label"><svg class="w-5 h-5" width="35" height="35" viewBox="0 0 35 35" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M6 12L17.8587 23.8587L29.7173 12" stroke="black" stroke-width="3"></path></svg><p class="ml-2 uppercase font-semibold">Table of Contents</p></span></button><div class="MuiCollapse-root MuiCollapse-hidden" style="min-height:0px"><div class="MuiCollapse-wrapper"><div class="MuiCollapse-wrapperInner"><ul class="flex flex-col border-l border-black"><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent screen-1425:border-black" data-active="true" data-testid="tocItem"><a class="screen-1425:font-bold" href="#using-dask-delayed-to-load-images">using dask.delayed to load images</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#make-your-life-easier-with-dask-image">make your life easier with dask-image</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#processing-data-with-dask-array-map-blocks">processing data with dask.array.map_blocks</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#further-reading">Further Reading</a></li></ul></div></div></div></div></div></div><div class="prose max-w-full App_content__2T0iD" role="main">
        <div id="documentation_options" data-url_root="../../"></div>
        <div class="section" id="using-dask-and-napari-to-process-view-large-datasets">

<p>Often in microscopy, multidimensional data is acquired and written to disk in many small files,
each of which contain a subset of one or more dimensions from the complete dataset.
For example, in a 5-dimensional experiment
(e.g. 3D z-stacks acquired for multiple channels at various moments over time),
each file on disk might be a 3D TIFF stack from a single channel at a single timepoint.
Data may also be stored in some proprietary format.
As the size of the dataset grows,
viewing arbitrary slices (in time, channel, z) of these datasets can become cumbersome.</p>
<p>Chunked file formats exist (such as <a class="reference external" href="https://support.hdfgroup.org/HDF5/" target="_blank" rel="noreferrer">hdf5</a> and <a class="reference external" href="https://zarr.readthedocs.io/en/stable/" target="_blank" rel="noreferrer">zarr</a>) that store data in a way that makes it easier to retrieve arbitrary subsets of the dataset, but they require either data duplication, or “committing” to a new file standard.</p>
<blockquote>
<div><p><strong>Note</strong>: This tutorial is not meant to promote a folder of TIFFs as a “good way” to store large datasets on disk;
but it is undoubtedly a common scenario in microscopy.
Chunked formats such as <code class="docutils literal notranslate"><span class="pre">hdf5</span></code> or <code class="docutils literal notranslate"><span class="pre">zarr</span></code> are superior in many ways,
but they do require the user to either duplicate their data
or go “all in” and delete the original data after conversion.
And while <code class="docutils literal notranslate"><span class="pre">napari</span></code> can easily handle something like a <code class="docutils literal notranslate"><span class="pre">zarr</span></code> store,
it can be a bit more limiting inasmuch as it requires programs that are capable of viewing it (i.e. you can’t necessarily just drag it into Fiji …)</p>
</div></blockquote>
<p>The first part of this tutorial demonstrates how to use <a class="reference external" href="https://docs.dask.org/en/latest/" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">Dask</span></code></a>
and <a class="reference external" href="https://docs.dask.org/en/latest/delayed.html" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">dask.delayed</span></code></a>
(or <a class="reference external" href="https://github.com/dask/dask-image" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">dask_image</span></code></a>) to feed <code class="docutils literal notranslate"><span class="pre">napari</span></code> image data “<a class="reference external" href="https://en.wikipedia.org/wiki/Lazy_evaluation" target="_blank" rel="noreferrer">lazily</a>”:
that is, the specific image file corresponding to the requested timepoint/channel
is only read from disk at the moment it is required
(based on the current position of the dimension sliders in <code class="docutils literal notranslate"><span class="pre">napari</span></code>).
Additionally, we will see that <em>any</em> function that takes a filepath
and returns a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array can be used to lazily read image data.
This can be useful if you have a proprietary format that is not immediately recognized by <code class="docutils literal notranslate"><span class="pre">napari</span></code>
(but for which you have at least some way of reading into a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array)</p>
<p>In some cases, data must be further processed prior to viewing,
such as a deskewing step for images acquired on a stage-scanning light sheet microscope.
Or perhaps you’d like to apply some basic image corrections or ratiometry prior to viewing.
The second part of this tutorial demonstrates the use of the <a class="reference external" href="https://docs.dask.org/en/latest/array-api.html#dask.array.map_blocks" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">dask.array.map_blocks</span></code></a> function
to describe an arbitrary sequence of functions in a declarative manner
that will be performed <em>on demand</em> as you explore the data (i.e. move the sliders) in <code class="docutils literal notranslate"><span class="pre">napari</span></code>.</p>
<div class="section" id="using-dask-delayed-to-load-images">
<h2>using dask.delayed to load images<a class="headerlink" href="#using-dask-delayed-to-load-images" title="Permalink to this headline">¶</a></h2>
<p>If you have a function that can take a filename and return a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array,
such as <code class="docutils literal notranslate"><span class="pre">skimage.io.imread</span></code>,
you can create a “lazy” version of that function by calling <code class="docutils literal notranslate"><span class="pre">dask.delayed</span></code> on the function itself.
This <em>new</em> function, when handed a filename,
will not actually read the file until explicitly asked with the <code class="docutils literal notranslate"><span class="pre">compute()</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">skimage.io</span> <span class="kn">import</span> <span class="n">imread</span>
<span class="kn">from</span> <span class="nn">dask</span> <span class="kn">import</span> <span class="n">delayed</span>

<span class="n">lazy_imread</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="n">imread</span><span class="p">)</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">lazy_imread</span><span class="p">(</span><span class="s1">'/path/to/file.tif'</span><span class="p">)</span>  <span class="c1"># doesn't actually read the file</span>
<span class="n">array</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>  <span class="c1"># *now* it reads.</span>
</pre></div>
</div>
<p>(If you have an unusual image format,
but you <em>do</em> have a python function that returns a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array,
simply substitute it for <code class="docutils literal notranslate"><span class="pre">skimage.io.imread</span></code> in the example above).</p>
<p>We can create a <a class="reference external" href="https://docs.dask.org/en/latest/array.html" target="_blank" rel="noreferrer">Dask array</a> of delayed file-readers
for <em>all</em> of the files in our multidimensional experiment using the <code class="docutils literal notranslate"><span class="pre">dask.array.from_delayed</span></code> function
and a <a class="reference external" href="https://docs.python.org/3/library/glob.html" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">glob</span></code></a> filename pattern
(<em>this example assumes that all files are of the same <code class="docutils literal notranslate"><span class="pre">shape</span></code> and <code class="docutils literal notranslate"><span class="pre">dtype</span></code>!</em>):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">skimage.io</span> <span class="kn">import</span> <span class="n">imread</span>
<span class="kn">from</span> <span class="nn">skimage.io.collection</span> <span class="kn">import</span> <span class="n">alphanumeric_key</span>
<span class="kn">from</span> <span class="nn">dask</span> <span class="kn">import</span> <span class="n">delayed</span>
<span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>
<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>

<span class="n">filenames</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">glob</span><span class="p">(</span><span class="s2">"/path/to/experiment/*.tif"</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">alphanumeric_key</span><span class="p">)</span>
<span class="c1"># read the first file to get the shape and dtype</span>
<span class="c1"># ASSUMES THAT ALL FILES SHARE THE SAME SHAPE/TYPE</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">imread</span><span class="p">(</span><span class="n">filenames</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="n">lazy_imread</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="n">imread</span><span class="p">)</span>  <span class="c1"># lazy reader</span>
<span class="n">lazy_arrays</span> <span class="o">=</span> <span class="p">[</span><span class="n">lazy_imread</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span> <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>
<span class="n">dask_arrays</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">da</span><span class="o">.</span><span class="n">from_delayed</span><span class="p">(</span><span class="n">delayed_reader</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">sample</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">delayed_reader</span> <span class="ow">in</span> <span class="n">lazy_arrays</span>
<span class="p">]</span>
<span class="c1"># Stack into one large dask.array</span>
<span class="n">stack</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">dask_arrays</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">stack</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># (nfiles, nz, ny, nx)</span>

<span class="c1"># in jupyter notebook the repr of a dask stack provides a useful visual:</span>
<span class="n">stack</span>
</pre></div>
</div>
<p><img alt="image: dask array html representation" src="../../_images/dask_repr.png"></p>
<p><em>No data has been read from disk yet!</em></p>
<p><code class="docutils literal notranslate"><span class="pre">napari</span></code> is capable of consuming Dask arrays,
so you can simply call <code class="docutils literal notranslate"><span class="pre">napari.view_image</span></code> on this <code class="docutils literal notranslate"><span class="pre">stack</span></code> and behind the scenes,
Dask will take care of reading the data from disk
and handing a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array to <code class="docutils literal notranslate"><span class="pre">napari</span></code> each time a new timepoint or channel is requested.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">napari</span>

<span class="c1"># specify contrast_limits and is_pyramid=False with big data</span>
<span class="c1"># to avoid unnecessary computations</span>
<span class="n">napari</span><span class="o">.</span><span class="n">view_image</span><span class="p">(</span><span class="n">stack</span><span class="p">,</span> <span class="n">contrast_limits</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2000</span><span class="p">],</span> <span class="n">multiscale</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p><em>Note: providing the</em> <code class="docutils literal notranslate"><span class="pre">contrast_limits</span></code> <em>and</em> <code class="docutils literal notranslate"><span class="pre">multiscale</span></code> <em>arguments prevents</em> <code class="docutils literal notranslate"><span class="pre">napari</span></code> <em>from trying to calculate the data min/max, which can take an extremely long time with big data.
See <a class="reference external" href="https://github.com/napari/napari/issues/736" target="_blank" rel="noreferrer">napari issue #736</a> for further discussion.</em></p>
</div>
<div class="section" id="make-your-life-easier-with-dask-image">
<h2>make your life easier with <code class="docutils literal notranslate"><span class="pre">dask-image</span></code><a class="headerlink" href="#make-your-life-easier-with-dask-image" title="Permalink to this headline">¶</a></h2>
<p>This pattern for creating a <code class="docutils literal notranslate"><span class="pre">dask.array</span></code> from image data
has been previously described in an <a class="reference external" href="https://blog.dask.org/2019/06/20/load-image-data" target="_blank" rel="noreferrer">excellent blog post</a> by John Kirkham.
It is a common-enough pattern that John created a useful library (<a class="reference external" href="https://github.com/dask/dask-image" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">dask-image</span></code></a>)
that does all this for you,
provided your image format can be read by the <a class="reference external" href="https://github.com/soft-matter/pims" target="_blank" rel="noreferrer"><code class="docutils literal notranslate"><span class="pre">pims</span></code></a> (Python Image Sequence) reader
(if not, see note above about providing your own reader function with <code class="docutils literal notranslate"><span class="pre">dask.delayed</span></code>).</p>
<p>Using <code class="docutils literal notranslate"><span class="pre">dask-image</span></code>, <em>all</em> of the above code can be simplified to 5 lines:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">napari</span>
<span class="kn">from</span> <span class="nn">dask_image.imread</span> <span class="kn">import</span> <span class="n">imread</span>

<span class="n">stack</span> <span class="o">=</span> <span class="n">imread</span><span class="p">(</span><span class="s2">"/path/to/experiment/*.tif"</span><span class="p">)</span>
<span class="n">napari</span><span class="o">.</span><span class="n">view_image</span><span class="p">(</span><span class="n">stack</span><span class="p">,</span> <span class="n">contrast_limits</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2000</span><span class="p">],</span> <span class="n">is_pyramid</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="image: mCherry-H2B showing chromosome separation during mitosis. Collected on a lattice light sheet microscope" src="../../_images/dask1.gif"></p>
<div class="section" id="side-note-regarding-higher-dimensional-datasets">
<h3><strong>side note regarding higher-dimensional datasets</strong><a class="headerlink" href="#side-note-regarding-higher-dimensional-datasets" title="Permalink to this headline">¶</a></h3>
<p>In the above example, it would be quite common to have a 5+ dimensional dataset
(e.g. different timepoints <em>and</em> channels represented among the 3D TIFF files in a folder).
A standard approach to deal with that sort of thing in <code class="docutils literal notranslate"><span class="pre">numpy</span></code> would be to <code class="docutils literal notranslate"><span class="pre">reshape</span></code> the array after instantiation.
With Dask, reshaping arrays can <em>sometimes</em> lead to unexpected read events if you’re not careful.</p>
<p>For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">dask_image.imread</span> <span class="kn">import</span> <span class="n">imread</span>
<span class="n">stack</span> <span class="o">=</span> <span class="n">imread</span><span class="p">(</span><span class="s1">'/path/to/experiment/*.tif'</span><span class="p">)</span>
<span class="n">stack</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># -&gt; something like (1200, 64, 256, 280)</span>
<span class="n">stack</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>  <span class="c1"># incurs a single file read</span>

<span class="c1"># if there were two channels in that folder you might try to do this:</span>
<span class="n">stack</span> <span class="o">=</span> <span class="n">stack</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">600</span><span class="p">,</span> <span class="mi">64</span><span class="p">,</span> <span class="mi">256</span><span class="p">,</span> <span class="mi">280</span><span class="p">)</span>
<span class="c1"># but now trying to read just the first timepoint in the first channel:</span>
<span class="n">stack</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>  <span class="c1"># incurs 600 read events!</span>
</pre></div>
</div>
<p>We will update this post as best-practices emerge,
but one possible solution to this is to avoid reshaping
Dask arrays altogether, by constructing multiple Dask arrays using <code class="docutils literal notranslate"><span class="pre">dask-image</span></code>, and then using
<code class="docutils literal notranslate"><span class="pre">da.stack</span></code> to combine them:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">dask_image.imread</span> <span class="kn">import</span> <span class="n">imread</span>
<span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>

<span class="c1"># # instead of this</span>
<span class="c1"># stack = imread('/path/to/experiment/*.tif')</span>
<span class="c1"># stack = stack.reshape(2, 600, 64, 256, 280)</span>
<span class="c1"># stack[0, 0].compute()  # incurs 600 read events!</span>

<span class="c1"># do something like this:</span>
<span class="n">file_pattern</span> <span class="o">=</span> <span class="s2">"/path/to/experiment/*ch</span><span class="si">{}</span><span class="s2">*.tif"</span>
<span class="n">channels</span> <span class="o">=</span> <span class="p">[</span><span class="n">imread</span><span class="p">(</span><span class="n">file_pattern</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nchannels</span><span class="p">)]</span>
<span class="n">stack</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">channels</span><span class="p">)</span>
<span class="n">stack</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># (2, 600, 64, 256, 280)</span>
<span class="n">stack</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>  <span class="c1"># incurs a single file read</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="processing-data-with-dask-array-map-blocks">
<h2>processing data with <code class="docutils literal notranslate"><span class="pre">dask.array.map_blocks</span></code><a class="headerlink" href="#processing-data-with-dask-array-map-blocks" title="Permalink to this headline">¶</a></h2>
<p>As previously mentioned,
sometimes it is desirable to process data prior to viewing.
We’ll take as an example
a series of TIFF files acquired on a lattice-light-sheet microscope.
A typical workflow might be to deskew, deconvolve, and perhaps crop
or apply some channel registration prior to viewing.</p>
<p>With <code class="docutils literal notranslate"><span class="pre">dask.array.map_blocks</span></code> we can apply any function that accepts a <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array
and returns a modified array to all the images in our <code class="docutils literal notranslate"><span class="pre">dask.array</span></code>.
It will be evaluated lazily, when requested (in this case, by <code class="docutils literal notranslate"><span class="pre">napari</span></code>);
we do not have to wait for it to process the entire dataset.</p>
<p>Here is an example of a script that will take a folder of raw tiff files,
and lazily read, deskew, deconvolve, crop,
and display them, <em>on demand</em> as you move the <code class="docutils literal notranslate"><span class="pre">napari</span></code> dimensions sliders around.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">napari</span>
<span class="kn">import</span> <span class="nn">pycudadecon</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>
<span class="kn">from</span> <span class="nn">skimage</span> <span class="kn">import</span> <span class="n">io</span>
<span class="kn">from</span> <span class="nn">dask_image.imread</span> <span class="kn">import</span> <span class="n">imread</span>

<span class="c1"># load stacks with dask_image, and psf with skimage</span>
<span class="n">stack</span> <span class="o">=</span> <span class="n">imread</span><span class="p">(</span><span class="s2">"/path/to/experiment/*.tif"</span><span class="p">)</span>
<span class="n">psf</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s2">"/path/to/psf.tif"</span><span class="p">)</span>

<span class="c1"># prepare some functions that accept a numpy array</span>
<span class="c1"># and return a processed array</span>

<span class="k">def</span> <span class="nf">last3dims</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
    <span class="c1"># this is just a wrapper because the pycudadecon function</span>
    <span class="c1"># expects ndims==3 but our blocks will have ndim==4</span>
    <span class="k">def</span> <span class="nf">func</span><span class="p">(</span><span class="n">array</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">f</span><span class="p">(</span><span class="n">array</span><span class="p">[</span><span class="mi">0</span><span class="p">])[</span><span class="kc">None</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">func</span>


<span class="k">def</span> <span class="nf">crop</span><span class="p">(</span><span class="n">array</span><span class="p">):</span>
    <span class="c1"># simple cropping function</span>
    <span class="k">return</span> <span class="n">array</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:,</span> <span class="mi">10</span><span class="p">:</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="p">:</span><span class="mi">500</span><span class="p">]</span>

<span class="c1"># https://docs.python.org/3.8/library/functools.html#functools.partial</span>
<span class="n">deskew</span> <span class="o">=</span> <span class="n">last3dims</span><span class="p">(</span><span class="n">partial</span><span class="p">(</span><span class="n">pycudadecon</span><span class="o">.</span><span class="n">deskew_gpu</span><span class="p">,</span> <span class="n">angle</span><span class="o">=</span><span class="mf">31.5</span><span class="p">))</span>
<span class="n">deconv</span> <span class="o">=</span> <span class="n">last3dims</span><span class="p">(</span><span class="n">partial</span><span class="p">(</span><span class="n">pycudadecon</span><span class="o">.</span><span class="n">decon</span><span class="p">,</span> <span class="n">psf</span><span class="o">=</span><span class="n">psf</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
<span class="c1"># note: this is done in two steps just as an example...</span>
<span class="c1"># in reality pycudadecon.decon also has a deskew argument</span>

<span class="c1"># map and chain those functions across all dask blocks</span>
<span class="n">deskewed</span> <span class="o">=</span> <span class="n">stack</span><span class="o">.</span><span class="n">map_blocks</span><span class="p">(</span><span class="n">deskew</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">"uint16"</span><span class="p">)</span>
<span class="n">deconvolved</span> <span class="o">=</span> <span class="n">deskewed</span><span class="o">.</span><span class="n">map_blocks</span><span class="p">(</span><span class="n">deconv</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">"float32"</span><span class="p">)</span>
<span class="n">cropped</span> <span class="o">=</span> <span class="n">deconvolved</span><span class="o">.</span><span class="n">map_blocks</span><span class="p">(</span><span class="n">crop</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">"float32"</span><span class="p">)</span>

<span class="c1"># put the resulting dask array into napari.</span>
<span class="c1"># (don't forget the contrast limits and is_pyramid==False !)</span>
<span class="n">v</span> <span class="o">=</span> <span class="n">napari</span><span class="o">.</span><span class="n">view_image</span><span class="p">(</span>
    <span class="n">cropped</span><span class="p">,</span>
    <span class="n">contrast_limits</span><span class="o">=</span><span class="p">[</span><span class="mi">90</span><span class="p">,</span> <span class="mi">1500</span><span class="p">],</span>
    <span class="n">is_pyramid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ndisplay</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">scale</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="p">)</span>

<span class="n">napari</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>Of course, the GUI isn’t as responsive as it would be if you had processed the data up front
and loaded the results into RAM and viewed them in <code class="docutils literal notranslate"><span class="pre">napari</span></code> (it’s doing a lot of work after all!),
but it’s surprisingly usable,
and allows you to preview the result of a relatively complex processing pipeline <em>on-the-fly</em>,
for arbitrary timepoints/channels, while storing <em>only</em> the raw data on disk.</p>
<p><img alt="image: same dataset, demonstrating on-the-fly read → deskew → deconvolve → crop" src="../../_images/dask2.gif"></p>
<p>This workflow is very much patterned after <a class="reference external" href="https://blog.dask.org/2019/08/09/image-itk" target="_blank" rel="noreferrer">another great post by John Kirkham, Matthew Rocklin, and Matthew McCormick</a>
that describes a similar image processing pipeline using <a class="reference external" href="https://itk.org/" target="_blank" rel="noreferrer">ITK</a>.
<code class="docutils literal notranslate"><span class="pre">napari</span></code> simply sits at the end of this lazy processing chain,
ready to show you the result on demand!</p>
</div>
<div class="section" id="further-reading">
<h2>Further Reading<a class="headerlink" href="#further-reading" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://docs.dask.org/en/latest/delayed.html" target="_blank" rel="noreferrer">Documentation on dask.delayed</a></p>
<p><a class="reference external" href="https://blog.dask.org/2019/06/20/load-image-data" target="_blank" rel="noreferrer">Dask working notes on dask-image</a></p>
<p><a class="reference external" href="https://blog.dask.org/2019/08/09/image-itk" target="_blank" rel="noreferrer">Dask working notes on image processing with <code class="docutils literal notranslate"><span class="pre">dask.array.map_blocks</span></code></a></p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./tutorials/applications"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>
      </div></div><div class="fresnel-container fresnel-greaterThanOrEqual-screen-1425 "><div class="fresnel-container fresnel-greaterThanOrEqual-screen-1425 sticky top-12"><ul class="flex flex-col border-l border-black"><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent screen-1425:border-black" data-active="true" data-testid="tocItem"><a class="screen-1425:font-bold" href="#using-dask-delayed-to-load-images">using dask.delayed to load images</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#make-your-life-easier-with-dask-image">make your life easier with dask-image</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#processing-data-with-dask-array-map-blocks">processing data with dask.array.map_blocks</a></li><li class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent" data-active="false" data-testid="tocItem"><a class="" href="#further-reading">Further Reading</a></li></ul></div><div class="fresnel-container fresnel-lessThan-screen-1425 "></div></div></div></main><div class="flex flex-row flex-wrap items-center justify-start w-full pt-6 pb-2 px-6 screen-495:px-12 bg-black"><a href="/" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center"><svg class="h-[1em] w-[1em] inline-block mr-1" width="30" height="30" viewBox="0 0 30 30" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Return to Home Page</title><path d="M27.8387 0H2.16127C0.967636 0 0 0.967636 0 2.16127V27.8387C0 29.0324 0.967636 30 2.16127 30H27.8387C29.0324 30 30 29.0324 30 27.8387V2.16127C30 0.967636 29.0324 0 27.8387 0Z" fill="#26283D"></path><path d="M27.7779 1.03139H2.22356C1.56514 1.03139 1.03139 1.56514 1.03139 2.22357V27.7779C1.03139 28.4363 1.56514 28.9701 2.22356 28.9701H27.7779C28.4363 28.9701 28.9701 28.4363 28.9701 27.7779V2.22357C28.9701 1.56514 28.4363 1.03139 27.7779 1.03139Z" fill="url(#paint0_radial)"></path><path d="M17.4655 27.3361C17.099 27.3361 16.7542 27.3201 16.4124 27.2897C16.0096 27.2645 15.61 27.202 15.2187 27.1028C14.4296 26.8759 13.6918 26.4986 13.0459 25.9918C12.1394 25.2971 11.3826 24.4265 10.8209 23.4322C10.5949 23.0338 10.3863 22.4775 9.83293 20.9899C9.71125 20.661 9.61999 20.4104 9.54321 20.2091C9.36359 19.7209 9.3042 19.5558 9.15789 19.2385C8.96705 18.7853 8.73274 18.3516 8.45823 17.9435C8.12277 17.4768 7.73384 17.0509 7.29937 16.6745C7.13279 16.5297 7.05601 16.4616 6.91405 16.3515C6.73732 16.2067 6.49396 16.0183 6.02752 15.6098C5.2395 14.9174 4.94254 14.5958 4.66876 14.2453C4.25597 13.7417 3.92726 13.1748 3.69532 12.5664C3.49242 12.0272 3.37416 11.4599 3.34476 10.8846C3.32978 10.3829 3.38289 9.88145 3.50265 9.39401C3.69467 8.45513 4.01235 7.54642 4.44713 6.69241C4.7481 6.05734 5.13833 5.46858 5.60599 4.94398C6.16001 4.34736 6.8007 3.83754 7.50652 3.43167C7.8302 3.24114 8.16941 3.07832 8.52052 2.94495C9.04221 2.75721 9.59283 2.66258 10.1473 2.66537C10.6436 2.66652 11.1376 2.73375 11.6161 2.86528C12.5756 3.10049 13.433 3.64024 14.0599 4.40366C14.7204 5.22066 14.987 6.2477 15.2028 7.63978C15.2578 8.00193 15.2796 8.18879 15.2984 8.35248C15.31 8.46547 15.323 8.57846 15.352 8.77836C15.5273 9.97199 15.6142 10.5167 15.8141 10.9324C16.1038 11.5408 16.6572 11.8957 17.0208 12.1304C17.9131 12.7098 18.8344 12.788 19.4442 12.8416C19.6847 12.8633 19.9585 12.8749 20.2569 12.8749C20.703 12.8749 21.0739 12.8503 21.2115 12.8402C21.4853 12.8199 21.7141 12.7938 21.9358 12.7677C22.272 12.7223 22.6106 12.6972 22.9498 12.6924C23.3746 12.6921 23.7976 12.7491 24.2071 12.8619C25.0966 13.067 25.8989 13.547 26.5002 14.2337C27.1101 14.958 27.2738 15.6721 27.4795 16.5804C27.6193 17.187 27.707 17.8044 27.7417 18.4259C27.7764 19.1313 27.8242 20.0946 27.4925 21.2303C27.1727 22.2747 26.6228 23.234 25.8831 24.0377C25.1073 24.908 24.1734 25.6233 23.1309 26.1453C22.5393 26.4354 21.9207 26.6669 21.2839 26.8363C20.4006 27.0835 19.4935 27.2365 18.578 27.2926C18.1912 27.3216 17.8175 27.3361 17.4655 27.3361Z" fill="#26283D"></path><path d="M17.4655 26.2265C17.1323 26.2265 16.8194 26.2134 16.5109 26.1859C16.1688 26.1658 15.8293 26.1139 15.4969 26.0309C14.8468 25.8412 14.2394 25.5279 13.7079 25.1082C12.924 24.5046 12.2698 23.7494 11.7842 22.8875C11.6016 22.5645 11.3873 21.9908 10.8701 20.6031C10.7484 20.2786 10.6586 20.0323 10.5804 19.8295C10.4008 19.3385 10.3269 19.1342 10.1618 18.7764C9.94359 18.2639 9.6768 17.7735 9.36504 17.3119C8.98218 16.7792 8.53831 16.293 8.04249 15.8633C7.84694 15.6852 7.74264 15.6026 7.59778 15.4882C7.41961 15.3433 7.19942 15.1738 6.75471 14.7842C6.03042 14.1497 5.76968 13.8672 5.53791 13.5703C5.19569 13.1552 4.92274 12.6875 4.7296 12.1854C4.41134 11.3736 4.36059 10.4814 4.58474 9.63882C4.75558 8.79295 5.04066 7.97424 5.43216 7.20521C5.68904 6.6581 6.02321 6.15074 6.42443 5.69869C6.89941 5.18449 7.44931 4.74497 8.05553 4.39498C8.32534 4.23625 8.6078 4.1001 8.90005 3.98793C9.29948 3.84394 9.72125 3.77185 10.1458 3.77499C10.5448 3.77727 10.9417 3.83182 11.3264 3.93723C12.0597 4.10871 12.7166 4.51577 13.1965 5.09609C13.6311 5.6422 13.8846 6.37663 14.1062 7.80492C14.1584 8.14389 14.1772 8.31048 14.1946 8.47127C14.2076 8.5944 14.2221 8.71753 14.254 8.93481C14.4336 10.1531 14.5321 10.8252 14.8146 11.4104C15.2492 12.2984 15.9734 12.7735 16.4182 13.056C17.5394 13.7803 18.6692 13.8817 19.3443 13.9425C19.6166 13.9672 19.9237 13.9788 20.2554 13.9788C20.7393 13.9788 21.142 13.9527 21.2941 13.9411C21.5838 13.9194 21.8387 13.8904 22.0604 13.8643C22.3529 13.8246 22.6474 13.8019 22.9425 13.7962C23.2691 13.7958 23.5942 13.8401 23.9087 13.9281C24.2796 14.028 25.0502 14.238 25.647 14.9421C26.0816 15.4592 26.2062 16.001 26.393 16.8252C26.5186 17.3689 26.5976 17.9223 26.6292 18.4795C26.661 19.14 26.7016 19.9614 26.422 20.9174C26.1469 21.8088 25.6746 22.6267 25.0401 23.3105C24.3582 24.0719 23.5378 24.6968 22.6224 25.1516C22.0928 25.4117 21.5387 25.6184 20.9681 25.7687C20.1555 25.9947 19.3214 26.1345 18.4795 26.1859C18.1333 26.2192 17.7929 26.2265 17.4655 26.2265Z" fill="#A59678"></path><path d="M9.21149 4.83535C10.0328 4.53405 10.8513 4.74409 11.0946 4.80783C11.6453 4.93454 12.1384 5.24019 12.4969 5.67697C12.8199 6.07533 13.0212 6.69098 13.2211 7.95558C13.3124 8.5437 13.295 8.56978 13.366 9.07823C13.5529 10.3472 13.6644 11.1062 14.0077 11.816C14.548 12.9314 15.4563 13.521 15.9329 13.8267C17.2366 14.6726 18.5403 14.7885 19.2646 14.8537C19.9624 14.9053 20.663 14.9053 21.3607 14.8537C22.4718 14.7711 22.9121 14.6045 23.6712 14.8117C24.0058 14.9015 24.549 15.0579 24.9546 15.536C25.2443 15.8851 25.337 16.2603 25.5109 17.0324C25.625 17.5246 25.6967 18.0257 25.7253 18.5302C25.7571 19.169 25.7904 19.86 25.5543 20.6697C25.3157 21.4363 24.906 22.1387 24.3563 22.7238C23.7531 23.3969 23.027 23.9485 22.2168 24.3491C21.736 24.5834 21.2327 24.7683 20.7146 24.901C19.9602 25.1106 19.1857 25.2397 18.4042 25.2863C17.7992 25.3373 17.1911 25.3373 16.5862 25.2863C16.2975 25.2698 16.0109 25.2267 15.7301 25.1574C15.1926 24.9997 14.6912 24.7384 14.254 24.3882C13.5654 23.8599 12.9908 23.1976 12.5649 22.4413C12.4201 22.1806 12.1811 21.5476 11.7088 20.2801C11.2916 19.1603 11.2323 18.943 10.9845 18.3969C10.7451 17.8345 10.451 17.297 10.1067 16.7919C9.68854 16.2019 9.20233 15.6632 8.65814 15.1869C8.19314 14.7639 8.16417 14.8044 7.35442 14.0961C6.64751 13.4761 6.43168 13.2342 6.25205 13.0053C5.96697 12.6629 5.73947 12.2764 5.57846 11.8609C5.45202 11.5217 5.37825 11.1651 5.35973 10.8035C5.35202 10.4734 5.39004 10.1439 5.47272 9.82425C5.62675 9.05475 5.88556 8.31001 6.24191 7.61082C6.46342 7.13752 6.75173 6.69847 7.09802 6.3071C7.32105 6.06667 7.56328 5.84479 7.82231 5.64366C8.03967 5.47474 8.26863 5.3213 8.50748 5.18446C8.73275 5.05002 8.96813 4.93329 9.21149 4.83535V4.83535Z" fill="url(#paint1_linear)"></path><defs><radialGradient id="paint0_radial" cx="0" cy="0" r="1" gradientUnits="userSpaceOnUse" gradientTransform="translate(15 15) scale(28.521)"><stop stop-color="#746B7F"></stop><stop offset="0.12" stop-color="#615B6F"></stop><stop offset="0.26" stop-color="#514C60"></stop><stop offset="0.37" stop-color="#4B475B"></stop><stop offset="0.44" stop-color="#4F4B5F"></stop><stop offset="0.51" stop-color="#5A566C"></stop><stop offset="0.59" stop-color="#6D6881"></stop><stop offset="0.67" stop-color="#87829E"></stop><stop offset="0.74" stop-color="#A9A2C3"></stop><stop offset="0.78" stop-color="#BCB5D9"></stop></radialGradient><linearGradient id="paint1_linear" x1="5.54949" y1="7.79624" x2="23.9884" y2="23.7393" gradientUnits="userSpaceOnUse"><stop offset="0.05" stop-color="#448491"></stop><stop offset="0.15" stop-color="#4A7986"></stop><stop offset="0.33" stop-color="#506E7B"></stop><stop offset="0.52" stop-color="#526A77"></stop><stop offset="0.6" stop-color="#56707C"></stop><stop offset="0.7" stop-color="#607F8B"></stop><stop offset="0.82" stop-color="#7099A3"></stop><stop offset="0.94" stop-color="#87BDC4"></stop><stop offset="1" stop-color="#94D1D6"></stop></linearGradient></defs></svg><span class="font-semibold text-[0.875em]">napari</span></a><a href="https://napari-hub.org" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center" target="_blank" rel="noreferrer"><svg class="h-[1em] w-[1em] inline-block mr-1" width="487" height="512" viewBox="0 0 487 512" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Visit napari hub</title><circle cx="243.036" cy="256" r="85.3333" fill="white" stroke="white" stroke-width="56.8889"></circle><circle cx="243.036" cy="42.6667" r="42.6667" fill="white"></circle><circle cx="243.036" cy="469.333" r="42.6667" fill="white"></circle><path d="M243.036 28.4444L243.036 142.222" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path><path d="M243.036 369.778L243.036 483.556" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path><circle cx="58.2838" cy="149.333" r="42.6667" transform="rotate(-60 58.2838 149.333)" fill="white"></circle><circle cx="427.788" cy="362.667" r="42.6667" transform="rotate(-60 427.788 362.667)" fill="white"></circle><path d="M45.967 142.222L144.501 199.111" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path><path d="M341.57 312.889L440.105 369.778" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path><circle cx="58.2838" cy="362.667" r="42.6667" transform="rotate(-120 58.2838 362.667)" fill="white"></circle><circle cx="427.788" cy="149.333" r="42.6667" transform="rotate(-120 427.788 149.333)" fill="white"></circle><path d="M45.967 369.778L144.501 312.889" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path><path d="M341.57 199.111L440.105 142.222" stroke="white" stroke-width="56.8889" stroke-linecap="round" stroke-linejoin="round"></path></svg><span class="font-semibold text-[0.875em]">napari hub</span></a><a href="https://github.com/napari/napari" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center" target="_blank" rel="noreferrer"><svg class="h-[1em] w-[1em] inline-block mr-1" width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Visit GitHub repository</title><path d="M7 0.173252C3.1325 0.173252 0 3.3075 0 7.17325C0 10.2667 2.0055 12.8899 4.78625 13.8145C5.13625 13.8804 5.26458 13.664 5.26458 13.4779C5.26458 13.3117 5.25875 12.8713 5.25583 12.2879C3.30867 12.7103 2.898 11.3488 2.898 11.3488C2.5795 10.5408 2.11925 10.325 2.11925 10.325C1.48517 9.891 2.16825 9.89975 2.16825 9.89975C2.87117 9.94875 3.24042 10.6208 3.24042 10.6208C3.86458 11.6912 4.879 11.382 5.27917 11.2029C5.34217 10.7503 5.52242 10.4417 5.7225 10.2667C4.16792 10.0917 2.534 9.48967 2.534 6.8075C2.534 6.04334 2.80525 5.41917 3.25442 4.92917C3.17567 4.75242 2.93942 4.04075 3.31567 3.0765C3.31567 3.0765 3.90192 2.88867 5.24067 3.794C5.80067 3.63825 6.39567 3.56125 6.99067 3.55775C7.58567 3.56125 8.18067 3.63825 8.74067 3.794C10.0707 2.88867 10.6569 3.0765 10.6569 3.0765C11.0332 4.04075 10.7969 4.75242 10.7269 4.92917C11.1732 5.41917 11.4444 6.04334 11.4444 6.8075C11.4444 9.49667 9.80817 10.0888 8.25067 10.2608C8.49567 10.4708 8.72317 10.9002 8.72317 11.5558C8.72317 12.4927 8.71442 13.2452 8.71442 13.4727C8.71442 13.6564 8.83692 13.8752 9.19567 13.8052C11.9962 12.887 14 10.262 14 7.17325C14 3.3075 10.8657 0.173252 7 0.173252Z" fill="white"></path></svg><span class="font-semibold text-[0.875em]">GitHub</span></a><a href="https://twitter.com/napari_imaging" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center" target="_blank" rel="noreferrer"><svg class="h-[1em] w-[1em] inline-block mr-1" width="15" height="13" viewBox="0 0 15 13" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Visit Twitter page</title><path d="M13.4581 3.44478C13.4676 3.57802 13.4676 3.7113 13.4676 3.84454C13.4676 7.9086 10.3744 12.5914 4.72082 12.5914C2.97906 12.5914 1.36105 12.0869 7.62939e-06 11.2113C0.247478 11.2398 0.485398 11.2494 0.742391 11.2494C2.17955 11.2494 3.50254 10.764 4.55901 9.93591C3.20749 9.90735 2.07487 9.0222 1.68464 7.80392C1.87501 7.83246 2.06535 7.8515 2.26524 7.8515C2.54125 7.8515 2.81728 7.81341 3.07425 7.74682C1.66562 7.46127 0.609119 6.22397 0.609119 4.72968V4.69163C1.01837 4.92006 1.49429 5.06282 1.9987 5.08183C1.17065 4.52979 0.628162 3.58755 0.628162 2.52155C0.628162 1.9505 0.780418 1.42702 1.04693 0.970164C2.56026 2.83564 4.83502 4.05389 7.38575 4.18717C7.33817 3.95874 7.30961 3.72082 7.30961 3.48287C7.30961 1.78869 8.68017 0.40863 10.3838 0.40863C11.269 0.40863 12.0685 0.779822 12.63 1.37944C13.3248 1.2462 13.9911 0.989207 14.5812 0.637058C14.3527 1.35091 13.8673 1.95052 13.2297 2.33121C13.8483 2.26462 14.4479 2.09326 14.9999 1.85534C14.5812 2.46445 14.0577 3.00694 13.4581 3.44478Z" fill="white"></path></svg><span class="font-semibold text-[0.875em]">Twitter</span></a><a href="https://forum.image.sc/tag/napari" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center" target="_blank" rel="noreferrer"><svg class="h-[1.25em] w-[1.25em] inline-block mr-1" width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Visit image.sc forum</title><path d="M12.9987 4.65361V2.01143C12.9987 2.00512 12.9936 2 12.9873 2H6.95296C6.94665 2 6.94153 2.00512 6.94153 2.01143V8.04571C6.94153 8.05203 6.94665 8.05714 6.95296 8.05714H10.9701" stroke="white" stroke-width="2.34286"></path><path d="M6.94153 15.3464V17.9886C6.94153 17.9949 6.94665 18 6.95296 18H12.9873C12.9936 18 12.9987 17.9949 12.9987 17.9886V11.9543C12.9987 11.948 12.9936 11.9429 12.9873 11.9429H8.9701" stroke="white" stroke-width="2.34286"></path><path d="M16.1145 9.9496L18.4027 8.62852C18.4082 8.62536 18.4101 8.61837 18.4069 8.6129L15.3898 3.38706C15.3866 3.38159 15.3796 3.37972 15.3742 3.38288L10.1483 6.40002C10.1429 6.40317 10.141 6.41016 10.1441 6.41563L12.1527 9.89458" stroke="white" stroke-width="2.34286"></path><path d="M3.82577 10.0503L1.53758 11.3714C1.53211 11.3745 1.53024 11.3815 1.5334 11.387L4.55054 16.6128C4.5537 16.6183 4.56069 16.6202 4.56615 16.617L9.792 13.5999C9.79746 13.5967 9.79934 13.5897 9.79618 13.5842L7.78761 10.1053" stroke="white" stroke-width="2.34286"></path><path d="M13.086 15.2962L15.3742 16.6173C15.3796 16.6205 15.3866 16.6186 15.3898 16.6131L18.4069 11.3873C18.4101 11.3818 18.4082 11.3748 18.4027 11.3717L13.1769 8.35456C13.1714 8.35141 13.1644 8.35328 13.1613 8.35875L11.1527 11.8377" stroke="white" stroke-width="2.34286"></path><path d="M6.85434 4.70397L4.56614 3.38288C4.56068 3.37973 4.55369 3.3816 4.55053 3.38707L1.53339 8.61291C1.53023 8.61838 1.53211 8.62537 1.53757 8.62852L6.76342 11.6457C6.76888 11.6488 6.77587 11.647 6.77903 11.6415L8.7876 8.16254" stroke="white" stroke-width="2.34286"></path></svg><span class="font-semibold text-[0.875em]">image.sc</span></a><a href="https://napari.zulipchat.com" class="text-white whitespace-nowrap mr-4 last:mr-0 mb-4 flex flex-row items-center" target="_blank" rel="noreferrer"><svg class="h-[1em] w-[1em] inline-block mr-1" width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg"><title>Visit Zulip chatroom</title><g clip-path="url(#clip0)"><path d="M15.178 2.39267C15.178 3.19867 14.816 3.91467 14.2647 4.34867L8.90869 9.13133C8.80935 9.21667 8.68002 9.07933 8.75202 8.96467L10.716 5.03133C10.7714 4.92133 10.7 4.786 10.5867 4.786H2.96802C1.78802 4.786 0.822021 3.70933 0.822021 2.39267C0.822021 1.07733 1.78802 3.09431e-07 2.96802 3.09431e-07H13.032C14.212 -0.000666357 15.178 1.076 15.178 2.39267ZM2.96802 16H13.032C14.212 16 15.178 14.9227 15.178 13.6067C15.178 12.2907 14.212 11.2133 13.032 11.2133H5.41335C5.30002 11.2133 5.22869 11.0787 5.28402 10.9687L7.24802 7.03533C7.32002 6.92067 7.19069 6.78333 7.09135 6.86867L1.73602 11.6507C1.18402 12.084 0.822688 12.8007 0.822688 13.6067C0.822688 14.9227 1.78802 16 2.96802 16Z" fill="white"></path></g><defs><clipPath id="clip0"><rect width="16" height="16" fill="white"></rect></clipPath></defs></svg><span class="font-semibold text-[0.875em]">Zulip</span></a></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"state":{"globalHeaders":{"https://napari-hub.org":{"level":1,"children":[],"href":"https://napari-hub.org","text":"napari hub"},"https://github.com/napari/napari":{"level":1,"children":[],"href":"https://github.com/napari/napari","text":"napari source code"},"/developers/index.html":{"level":1,"children":["/developers/benchmarks.html","/developers/contributing.html","/developers/core_dev_guide.html","/developers/profiling.html","/developers/release.html","/developers/testing.html","/developers/docs.html"],"href":"/developers/index.html","text":"Developer resources"},"/developers/docs.html":{"level":2,"children":[],"href":"/developers/docs.html","text":"Organization of Documentation for napari"},"/developers/testing.html":{"level":2,"children":[],"href":"/developers/testing.html","text":"Testing"},"/developers/release.html":{"level":2,"children":[],"href":"/developers/release.html","text":"Release Guide"},"/developers/profiling.html":{"level":2,"children":[],"href":"/developers/profiling.html","text":"Profiling"},"/developers/core_dev_guide.html":{"level":2,"children":[],"href":"/developers/core_dev_guide.html","text":"Core Developer Guide"},"/developers/contributing.html":{"level":2,"children":[],"href":"/developers/contributing.html","text":"Contributing Guide"},"/developers/benchmarks.html":{"level":2,"children":[],"href":"/developers/benchmarks.html","text":"Benchmarks"},"/guides/stable/index.html":{"level":1,"children":["/guides/stable/event_loop.html","/guides/stable/threading.html","/guides/stable/perfmon.html","/guides/stable/rendering.html","/guides/stable/docker.html","/guides/stable/performance.html","/guides/stable/magicgui.html","/guides/stable/translations.html","/guides/stable/rendering-explanation.html","/guides/stable/connecting_events.html","/guides/stable/3D_interactivity.html","/guides/stable/events_reference.html"],"href":"/guides/stable/index.html","text":"Developer guides"},"/guides/stable/events_reference.html":{"level":2,"children":[],"href":"/guides/stable/events_reference.html","text":"Events Reference"},"/guides/stable/3D_interactivity.html":{"level":2,"children":[],"href":"/guides/stable/3D_interactivity.html","text":"3D interactivity"},"/guides/stable/connecting_events.html":{"level":2,"children":[],"href":"/guides/stable/connecting_events.html","text":"Hooking up your own events"},"/guides/stable/rendering-explanation.html":{"level":2,"children":[],"href":"/guides/stable/rendering-explanation.html","text":"Rendering in napari"},"/guides/stable/translations.html":{"level":2,"children":[],"href":"/guides/stable/translations.html","text":"Translations"},"/guides/stable/magicgui.html":{"level":2,"children":[],"href":"/guides/stable/magicgui.html","text":"Using magicgui in napari"},"/guides/stable/performance.html":{"level":2,"children":[],"href":"/guides/stable/performance.html","text":"Napari Performance"},"/guides/stable/docker.html":{"level":2,"children":[],"href":"/guides/stable/docker.html","text":"Napari in Docker (WIP)"},"/guides/stable/rendering.html":{"level":2,"children":[],"href":"/guides/stable/rendering.html","text":"Asynchronous Rendering"},"/guides/stable/perfmon.html":{"level":2,"children":[],"href":"/guides/stable/perfmon.html","text":"Performance Monitoring"},"/guides/stable/threading.html":{"level":2,"children":[],"href":"/guides/stable/threading.html","text":"Multithreading in napari"},"/guides/stable/event_loop.html":{"level":2,"children":[],"href":"/guides/stable/event_loop.html","text":"An Introduction to the Event Loop in napari"},"/roadmaps/index.html":{"level":1,"children":["/roadmaps/0_4.html","/roadmaps/0_3_retrospective.html","/roadmaps/0_3.html"],"href":"/roadmaps/index.html","text":"Roadmaps"},"/roadmaps/0_3.html":{"level":2,"children":[],"href":"/roadmaps/0_3.html","text":"Roadmap 0.3"},"/roadmaps/0_3_retrospective.html":{"level":2,"children":[],"href":"/roadmaps/0_3_retrospective.html","text":"Roadmap 0.3 Retrospective"},"/roadmaps/0_4.html":{"level":2,"children":[],"href":"/roadmaps/0_4.html","text":"Roadmap 0.4"},"/api/stable/index.html":{"level":1,"children":["/api/stable/napari.layers.html","/api/stable/napari.view_layers.html","/api/stable/napari.types.html","/api/stable/napari.utils.html","/api/stable/napari.plugins.html","/api/stable/napari.components.html","/api/stable/napari.qt.threading.html","/api/stable/napari.utils.perf.html","/api/stable/napari.html"],"href":"/api/stable/index.html","text":"API reference"},"/api/stable/napari.html":{"level":2,"children":["/api/stable/napari.Viewer.html"],"href":"/api/stable/napari.html","text":"napari"},"/api/stable/napari.Viewer.html":{"level":3,"children":[],"href":"/api/stable/napari.Viewer.html","text":"napari.Viewer"},"/api/stable/napari.utils.perf.html":{"level":2,"children":["/api/stable/napari.utils.perf.PerfEvent.html"],"href":"/api/stable/napari.utils.perf.html","text":"napari.utils.perf"},"/api/stable/napari.utils.perf.PerfEvent.html":{"level":3,"children":[],"href":"/api/stable/napari.utils.perf.PerfEvent.html","text":"napari.utils.perf.PerfEvent"},"/api/stable/napari.qt.threading.html":{"level":2,"children":["/api/stable/napari.qt.threading.FunctionWorker.html","/api/stable/napari.qt.threading.GeneratorWorker.html","/api/stable/napari.qt.threading.GeneratorWorkerSignals.html","/api/stable/napari.qt.threading.WorkerBase.html","/api/stable/napari.qt.threading.WorkerBaseSignals.html"],"href":"/api/stable/napari.qt.threading.html","text":"napari.qt.threading"},"/api/stable/napari.qt.threading.WorkerBaseSignals.html":{"level":3,"children":[],"href":"/api/stable/napari.qt.threading.WorkerBaseSignals.html","text":"napari.qt.threading.WorkerBaseSignals"},"/api/stable/napari.qt.threading.WorkerBase.html":{"level":3,"children":[],"href":"/api/stable/napari.qt.threading.WorkerBase.html","text":"napari.qt.threading.WorkerBase"},"/api/stable/napari.qt.threading.GeneratorWorkerSignals.html":{"level":3,"children":[],"href":"/api/stable/napari.qt.threading.GeneratorWorkerSignals.html","text":"napari.qt.threading.GeneratorWorkerSignals"},"/api/stable/napari.qt.threading.GeneratorWorker.html":{"level":3,"children":[],"href":"/api/stable/napari.qt.threading.GeneratorWorker.html","text":"napari.qt.threading.GeneratorWorker"},"/api/stable/napari.qt.threading.FunctionWorker.html":{"level":3,"children":[],"href":"/api/stable/napari.qt.threading.FunctionWorker.html","text":"napari.qt.threading.FunctionWorker"},"/api/stable/napari.components.html":{"level":2,"children":["/api/stable/napari.components.Camera.html","/api/stable/napari.components.Dims.html","/api/stable/napari.components.LayerList.html","/api/stable/napari.components.ViewerModel.html"],"href":"/api/stable/napari.components.html","text":"napari.components"},"/api/stable/napari.components.ViewerModel.html":{"level":3,"children":[],"href":"/api/stable/napari.components.ViewerModel.html","text":"napari.components.ViewerModel"},"/api/stable/napari.components.LayerList.html":{"level":3,"children":[],"href":"/api/stable/napari.components.LayerList.html","text":"napari.components.LayerList"},"/api/stable/napari.components.Dims.html":{"level":3,"children":[],"href":"/api/stable/napari.components.Dims.html","text":"napari.components.Dims"},"/api/stable/napari.components.Camera.html":{"level":3,"children":[],"href":"/api/stable/napari.components.Camera.html","text":"napari.components.Camera"},"/api/stable/napari.plugins.html":{"level":2,"children":["/api/stable/napari.plugins.NapariPluginManager.html"],"href":"/api/stable/napari.plugins.html","text":"napari.plugins"},"/api/stable/napari.plugins.NapariPluginManager.html":{"level":3,"children":[],"href":"/api/stable/napari.plugins.NapariPluginManager.html","text":"napari.plugins.NapariPluginManager"},"/api/stable/napari.utils.html":{"level":2,"children":["/api/stable/napari.utils.Colormap.html","/api/stable/napari.utils.nbscreenshot.html","/api/stable/napari.utils.progress.html"],"href":"/api/stable/napari.utils.html","text":"napari.utils"},"/api/stable/napari.utils.progress.html":{"level":3,"children":[],"href":"/api/stable/napari.utils.progress.html","text":"napari.utils.progress"},"/api/stable/napari.utils.nbscreenshot.html":{"level":3,"children":[],"href":"/api/stable/napari.utils.nbscreenshot.html","text":"napari.utils.nbscreenshot"},"/api/stable/napari.utils.Colormap.html":{"level":3,"children":[],"href":"/api/stable/napari.utils.Colormap.html","text":"napari.utils.Colormap"},"/api/stable/napari.types.html":{"level":2,"children":["/api/stable/napari.types.ArrayBase.html","/api/stable/napari.types.Path.html","/api/stable/napari.types.SampleDict.html","/api/stable/napari.types.TracebackType.html","/api/stable/napari.types.TypedDict.html","/api/stable/napari.types.partial.html"],"href":"/api/stable/napari.types.html","text":"napari.types"},"/api/stable/napari.types.partial.html":{"level":3,"children":[],"href":"/api/stable/napari.types.partial.html","text":"napari.types.partial"},"/api/stable/napari.types.TypedDict.html":{"level":3,"children":[],"href":"/api/stable/napari.types.TypedDict.html","text":"napari.types.TypedDict"},"/api/stable/napari.types.TracebackType.html":{"level":3,"children":[],"href":"/api/stable/napari.types.TracebackType.html","text":"napari.types.TracebackType"},"/api/stable/napari.types.SampleDict.html":{"level":3,"children":[],"href":"/api/stable/napari.types.SampleDict.html","text":"napari.types.SampleDict"},"/api/stable/napari.types.Path.html":{"level":3,"children":[],"href":"/api/stable/napari.types.Path.html","text":"napari.types.Path"},"/api/stable/napari.types.ArrayBase.html":{"level":3,"children":[],"href":"/api/stable/napari.types.ArrayBase.html","text":"napari.types.ArrayBase"},"/api/stable/napari.view_layers.html":{"level":2,"children":["/api/stable/napari.view_layers.Viewer.html"],"href":"/api/stable/napari.view_layers.html","text":"napari.view_layers"},"/api/stable/napari.view_layers.Viewer.html":{"level":3,"children":[],"href":"/api/stable/napari.view_layers.Viewer.html","text":"napari.view_layers.Viewer"},"/api/stable/napari.layers.html":{"level":2,"children":["/api/stable/napari.layers.Image.html","/api/stable/napari.layers.Labels.html","/api/stable/napari.layers.Layer.html","/api/stable/napari.layers.Points.html","/api/stable/napari.layers.Shapes.html","/api/stable/napari.layers.Surface.html","/api/stable/napari.layers.Tracks.html","/api/stable/napari.layers.Vectors.html"],"href":"/api/stable/napari.layers.html","text":"napari.layers"},"/api/stable/napari.layers.Vectors.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Vectors.html","text":"napari.layers.Vectors"},"/api/stable/napari.layers.Tracks.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Tracks.html","text":"napari.layers.Tracks"},"/api/stable/napari.layers.Surface.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Surface.html","text":"napari.layers.Surface"},"/api/stable/napari.layers.Shapes.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Shapes.html","text":"napari.layers.Shapes"},"/api/stable/napari.layers.Points.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Points.html","text":"napari.layers.Points"},"/api/stable/napari.layers.Layer.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Layer.html","text":"napari.layers.Layer"},"/api/stable/napari.layers.Labels.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Labels.html","text":"napari.layers.Labels"},"/api/stable/napari.layers.Image.html":{"level":3,"children":[],"href":"/api/stable/napari.layers.Image.html","text":"napari.layers.Image"},"/release/index.html":{"level":1,"children":["/release/release_0_4_12.html","/release/release_0_4_11.html","/release/release_0_4_10.html","/release/release_0_4_9.html","/release/release_0_4_8.html","/release/release_0_4_7.html","/release/release_0_4_6.html","/release/release_0_4_5.html","/release/release_0_4_4.html","/release/release_0_4_3.html","/release/release_0_4_2.html","/release/release_0_4_1.html","/release/release_0_4_0.html","/release/release_0_3_8.html","/release/release_0_3_7.html","/release/release_0_3_6.html","/release/release_0_3_5.html","/release/release_0_3_4.html","/release/release_0_3_3.html","/release/release_0_3_2.html","/release/release_0_3_1.html","/release/release_0_3_0.html","/release/release_0_2_12.html","/release/release_0_2_11.html","/release/release_0_2_10.html","/release/release_0_2_9.html","/release/release_0_2_8.html","/release/release_0_2_7.html","/release/release_0_2_6.html","/release/release_0_2_5.html","/release/release_0_2_4.html","/release/release_0_2_3.html","/release/release_0_2_1.html","/release/release_0_2_0.html","/release/release_0_1_5.html","/release/release_0_1_3.html","/release/release_0_1_0.html"],"href":"/release/index.html","text":"Release notes"},"/release/release_0_1_0.html":{"level":2,"children":[],"href":"/release/release_0_1_0.html","text":"napari 0.1.0"},"/release/release_0_1_3.html":{"level":2,"children":[],"href":"/release/release_0_1_3.html","text":"napari 0.1.3"},"/release/release_0_1_5.html":{"level":2,"children":[],"href":"/release/release_0_1_5.html","text":"napari 0.1.5"},"/release/release_0_2_0.html":{"level":2,"children":[],"href":"/release/release_0_2_0.html","text":"napari 0.2.0"},"/release/release_0_2_1.html":{"level":2,"children":[],"href":"/release/release_0_2_1.html","text":"napari 0.2.1"},"/release/release_0_2_3.html":{"level":2,"children":[],"href":"/release/release_0_2_3.html","text":"napari 0.2.3"},"/release/release_0_2_4.html":{"level":2,"children":[],"href":"/release/release_0_2_4.html","text":"napari 0.2.4"},"/release/release_0_2_5.html":{"level":2,"children":[],"href":"/release/release_0_2_5.html","text":"napari 0.2.5"},"/release/release_0_2_6.html":{"level":2,"children":[],"href":"/release/release_0_2_6.html","text":"napari 0.2.6"},"/release/release_0_2_7.html":{"level":2,"children":[],"href":"/release/release_0_2_7.html","text":"napari 0.2.7"},"/release/release_0_2_8.html":{"level":2,"children":[],"href":"/release/release_0_2_8.html","text":"napari 0.2.8"},"/release/release_0_2_9.html":{"level":2,"children":[],"href":"/release/release_0_2_9.html","text":"napari 0.2.9"},"/release/release_0_2_10.html":{"level":2,"children":[],"href":"/release/release_0_2_10.html","text":"napari 0.2.10"},"/release/release_0_2_11.html":{"level":2,"children":[],"href":"/release/release_0_2_11.html","text":"napari 0.2.11"},"/release/release_0_2_12.html":{"level":2,"children":[],"href":"/release/release_0_2_12.html","text":"napari 0.2.12"},"/release/release_0_3_0.html":{"level":2,"children":[],"href":"/release/release_0_3_0.html","text":"napari 0.3.0"},"/release/release_0_3_1.html":{"level":2,"children":[],"href":"/release/release_0_3_1.html","text":"napari 0.3.1"},"/release/release_0_3_2.html":{"level":2,"children":[],"href":"/release/release_0_3_2.html","text":"napari 0.3.2"},"/release/release_0_3_3.html":{"level":2,"children":[],"href":"/release/release_0_3_3.html","text":"napari 0.3.3"},"/release/release_0_3_4.html":{"level":2,"children":[],"href":"/release/release_0_3_4.html","text":"napari 0.3.4"},"/release/release_0_3_5.html":{"level":2,"children":[],"href":"/release/release_0_3_5.html","text":"napari 0.3.5"},"/release/release_0_3_6.html":{"level":2,"children":[],"href":"/release/release_0_3_6.html","text":"napari 0.3.6"},"/release/release_0_3_7.html":{"level":2,"children":[],"href":"/release/release_0_3_7.html","text":"napari 0.3.7"},"/release/release_0_3_8.html":{"level":2,"children":[],"href":"/release/release_0_3_8.html","text":"napari 0.3.8"},"/release/release_0_4_0.html":{"level":2,"children":[],"href":"/release/release_0_4_0.html","text":"napari 0.4.0"},"/release/release_0_4_1.html":{"level":2,"children":[],"href":"/release/release_0_4_1.html","text":"napari 0.4.1"},"/release/release_0_4_2.html":{"level":2,"children":[],"href":"/release/release_0_4_2.html","text":"napari 0.4.2"},"/release/release_0_4_3.html":{"level":2,"children":[],"href":"/release/release_0_4_3.html","text":"napari 0.4.3"},"/release/release_0_4_4.html":{"level":2,"children":[],"href":"/release/release_0_4_4.html","text":"napari 0.4.4"},"/release/release_0_4_5.html":{"level":2,"children":[],"href":"/release/release_0_4_5.html","text":"napari 0.4.5"},"/release/release_0_4_6.html":{"level":2,"children":[],"href":"/release/release_0_4_6.html","text":"napari 0.4.6"},"/release/release_0_4_7.html":{"level":2,"children":[],"href":"/release/release_0_4_7.html","text":"napari 0.4.7"},"/release/release_0_4_8.html":{"level":2,"children":[],"href":"/release/release_0_4_8.html","text":"napari 0.4.8"},"/release/release_0_4_9.html":{"level":2,"children":[],"href":"/release/release_0_4_9.html","text":"napari 0.4.9"},"/release/release_0_4_10.html":{"level":2,"children":[],"href":"/release/release_0_4_10.html","text":"napari 0.4.10"},"/release/release_0_4_11.html":{"level":2,"children":[],"href":"/release/release_0_4_11.html","text":"napari 0.4.11"},"/release/release_0_4_12.html":{"level":2,"children":[],"href":"/release/release_0_4_12.html","text":"napari 0.4.12"},"/plugins/stable/index.html":{"level":1,"children":["/plugins/stable/for_napari_developers.html","/plugins/stable/for_plugin_developers.html","/plugins/stable/hook_specifications.html","/plugins/stable/find-and-install-plugin.html"],"href":"/plugins/stable/index.html","text":"Plugins"},"/plugins/stable/find-and-install-plugin.html":{"level":2,"children":[],"href":"/plugins/stable/find-and-install-plugin.html","text":"Finding and installing a napari plugin"},"/plugins/stable/hook_specifications.html":{"level":2,"children":[],"href":"/plugins/stable/hook_specifications.html","text":"napari hook specification reference"},"/plugins/stable/for_plugin_developers.html":{"level":2,"children":[],"href":"/plugins/stable/for_plugin_developers.html","text":"creating a napari plugin"},"/plugins/stable/for_napari_developers.html":{"level":2,"children":[],"href":"/plugins/stable/for_napari_developers.html","text":"napari plugin architecture"},"/tutorials/index.html":{"level":1,"children":["/tutorials/fundamentals/index.html","/tutorials/applications/index.html","/tutorials/further-resources/napari-workshops.html"],"href":"/tutorials/index.html","text":"Tutorials"},"/tutorials/further-resources/napari-workshops.html":{"level":2,"children":[],"href":"/tutorials/further-resources/napari-workshops.html","text":"napari workshops"},"/tutorials/applications/index.html":{"level":2,"children":["/tutorials/applications/annotate_segmentation.html","/tutorials/applications/annotate_points.html","/tutorials/applications/dask.html","/tutorials/applications/napari_imageJ.html","/tutorials/applications/cell_tracking.html"],"href":"/tutorials/applications/index.html","text":"Applications"},"/tutorials/applications/cell_tracking.html":{"level":3,"children":[],"href":"/tutorials/applications/cell_tracking.html","text":"single cell tracking with napari"},"/tutorials/applications/napari_imageJ.html":{"level":3,"children":[],"href":"/tutorials/applications/napari_imageJ.html","text":"Napari + ImageJ How-to-Guide"},"/tutorials/applications/dask.html":{"level":3,"children":[],"href":"/tutorials/applications/dask.html","text":"using dask and napari to process \u0026 view large datasets"},"/tutorials/applications/annotate_points.html":{"level":3,"children":[],"href":"/tutorials/applications/annotate_points.html","text":"annotating videos with napari"},"/tutorials/applications/annotate_segmentation.html":{"level":3,"children":[],"href":"/tutorials/applications/annotate_segmentation.html","text":"annotating segmentation with text and bounding boxes"},"/tutorials/fundamentals/index.html":{"level":2,"children":["/tutorials/fundamentals/installation.html","/tutorials/fundamentals/getting_started.html","/tutorials/fundamentals/viewer.html","/tutorials/fundamentals/image.html","/tutorials/fundamentals/labels.html","/tutorials/fundamentals/points.html","/tutorials/fundamentals/shapes.html","/tutorials/fundamentals/surface.html","/tutorials/fundamentals/tracks.html","/tutorials/fundamentals/vectors.html"],"href":"/tutorials/fundamentals/index.html","text":"Fundamentals"},"/tutorials/fundamentals/vectors.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/vectors.html","text":"vectors layer tutorial"},"/tutorials/fundamentals/tracks.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/tracks.html","text":"tracks layer tutorial"},"/tutorials/fundamentals/surface.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/surface.html","text":"surface layer tutorial"},"/tutorials/fundamentals/shapes.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/shapes.html","text":"shapes layer tutorial"},"/tutorials/fundamentals/points.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/points.html","text":"points layer tutorial"},"/tutorials/fundamentals/labels.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/labels.html","text":"labels layer tutorial"},"/tutorials/fundamentals/image.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/image.html","text":"image layer tutorial"},"/tutorials/fundamentals/viewer.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/viewer.html","text":"napari viewer tutorial"},"/tutorials/fundamentals/getting_started.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/getting_started.html","text":"getting started with napari"},"/tutorials/fundamentals/installation.html":{"level":3,"children":[],"href":"/tutorials/fundamentals/installation.html","text":"napari installation tutorial"},"/community/index.html":{"level":1,"children":["/community/mission_and_values.html","/community/team.html","/community/code_of_conduct.html","/community/code_of_conduct_reporting.html","/community/governance.html"],"href":"/community/index.html","text":"Community"},"/community/governance.html":{"level":2,"children":[],"href":"/community/governance.html","text":"Governance Model"},"/community/code_of_conduct_reporting.html":{"level":2,"children":[],"href":"/community/code_of_conduct_reporting.html","text":"Handling Code of Conduct Reports"},"/community/code_of_conduct.html":{"level":2,"children":[],"href":"/community/code_of_conduct.html","text":"Code of Conduct"},"/community/team.html":{"level":2,"children":[],"href":"/community/team.html","text":"About the Project and Team"},"/community/mission_and_values.html":{"level":2,"children":[],"href":"/community/mission_and_values.html","text":"Mission and Values"}},"rootGlobalHeaders":["/community/index.html","/tutorials/index.html","/plugins/stable/index.html","/release/index.html","/api/stable/index.html","/roadmaps/index.html","/guides/stable/index.html","/developers/index.html","https://github.com/napari/napari","https://napari-hub.org"],"pageTitle":"using dask and napari to process \u0026 view large datasets","pageFrontMatter":{},"previewImage":"../../_images/dask_repr.png","previewDescription":"Often in microscopy, multidimensional data is acquired and written to disk in many small files,\neach of which contain a subset of one or more dimensions from the complete dataset.\nFor example, in a...","appScripts":[{"src":"/_static/documentation_options.js"},{"src":"/_static/jquery.js"},{"src":"/_static/underscore.js"},{"src":"/_static/doctools.js"},{"src":"/_static/togglebutton.js"},{"src":"/_static/clipboard.min.js"},{"src":"/_static/copybutton.js"},{"children":"var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';"},{"src":"https://unpkg.com/thebe@0.5.1/lib/index.js"},{"children":"\n        const thebe_selector = \".thebe\"\n        const thebe_selector_input = \"pre\"\n        const thebe_selector_output = \".output\"\n    "},{"src":"/_static/sphinx-thebe.js"}],"appStyleSheets":[{"href":"/_static/pygments.css","rel":"stylesheet"},{"href":"/_static/togglebutton.css","rel":"stylesheet","type":"text/css"},{"href":"/_static/copybutton.css","rel":"stylesheet","type":"text/css"},{"href":"/_static/mystnb.css","rel":"stylesheet","type":"text/css"},{"href":"/_static/sphinx-thebe.css","rel":"stylesheet","type":"text/css"},{"href":"/_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css","rel":"stylesheet","type":"text/css"},{"href":"/_static/panels-variables.06eb56fa6e07937060861dad626602ad.css","rel":"stylesheet","type":"text/css"}],"pageBodyHtml":"\n        \u003cdiv id=\"documentation_options\" data-url_root=\"../../\"\u003e\u003c/div\u003e\n        \u003cdiv class=\"section\" id=\"using-dask-and-napari-to-process-view-large-datasets\"\u003e\n\n\u003cp\u003eOften in microscopy, multidimensional data is acquired and written to disk in many small files,\neach of which contain a subset of one or more dimensions from the complete dataset.\nFor example, in a 5-dimensional experiment\n(e.g. 3D z-stacks acquired for multiple channels at various moments over time),\neach file on disk might be a 3D TIFF stack from a single channel at a single timepoint.\nData may also be stored in some proprietary format.\nAs the size of the dataset grows,\nviewing arbitrary slices (in time, channel, z) of these datasets can become cumbersome.\u003c/p\u003e\n\u003cp\u003eChunked file formats exist (such as \u003ca class=\"reference external\" href=\"https://support.hdfgroup.org/HDF5/\" target=\"_blank\" rel=\"noreferrer\"\u003ehdf5\u003c/a\u003e and \u003ca class=\"reference external\" href=\"https://zarr.readthedocs.io/en/stable/\" target=\"_blank\" rel=\"noreferrer\"\u003ezarr\u003c/a\u003e) that store data in a way that makes it easier to retrieve arbitrary subsets of the dataset, but they require either data duplication, or “committing” to a new file standard.\u003c/p\u003e\n\u003cblockquote\u003e\n\u003cdiv\u003e\u003cp\u003e\u003cstrong\u003eNote\u003c/strong\u003e: This tutorial is not meant to promote a folder of TIFFs as a “good way” to store large datasets on disk;\nbut it is undoubtedly a common scenario in microscopy.\nChunked formats such as \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003ehdf5\u003c/span\u003e\u003c/code\u003e or \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003ezarr\u003c/span\u003e\u003c/code\u003e are superior in many ways,\nbut they do require the user to either duplicate their data\nor go “all in” and delete the original data after conversion.\nAnd while \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e can easily handle something like a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003ezarr\u003c/span\u003e\u003c/code\u003e store,\nit can be a bit more limiting inasmuch as it requires programs that are capable of viewing it (i.e. you can’t necessarily just drag it into Fiji …)\u003c/p\u003e\n\u003c/div\u003e\u003c/blockquote\u003e\n\u003cp\u003eThe first part of this tutorial demonstrates how to use \u003ca class=\"reference external\" href=\"https://docs.dask.org/en/latest/\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eDask\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e\nand \u003ca class=\"reference external\" href=\"https://docs.dask.org/en/latest/delayed.html\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.delayed\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e\n(or \u003ca class=\"reference external\" href=\"https://github.com/dask/dask-image\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask_image\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e) to feed \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e image data “\u003ca class=\"reference external\" href=\"https://en.wikipedia.org/wiki/Lazy_evaluation\" target=\"_blank\" rel=\"noreferrer\"\u003elazily\u003c/a\u003e”:\nthat is, the specific image file corresponding to the requested timepoint/channel\nis only read from disk at the moment it is required\n(based on the current position of the dimension sliders in \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e).\nAdditionally, we will see that \u003cem\u003eany\u003c/em\u003e function that takes a filepath\nand returns a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array can be used to lazily read image data.\nThis can be useful if you have a proprietary format that is not immediately recognized by \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e\n(but for which you have at least some way of reading into a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array)\u003c/p\u003e\n\u003cp\u003eIn some cases, data must be further processed prior to viewing,\nsuch as a deskewing step for images acquired on a stage-scanning light sheet microscope.\nOr perhaps you’d like to apply some basic image corrections or ratiometry prior to viewing.\nThe second part of this tutorial demonstrates the use of the \u003ca class=\"reference external\" href=\"https://docs.dask.org/en/latest/array-api.html#dask.array.map_blocks\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array.map_blocks\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e function\nto describe an arbitrary sequence of functions in a declarative manner\nthat will be performed \u003cem\u003eon demand\u003c/em\u003e as you explore the data (i.e. move the sliders) in \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e.\u003c/p\u003e\n\u003cdiv class=\"section\" id=\"using-dask-delayed-to-load-images\"\u003e\n\u003ch2\u003eusing dask.delayed to load images\u003ca class=\"headerlink\" href=\"#using-dask-delayed-to-load-images\" title=\"Permalink to this headline\"\u003e¶\u003c/a\u003e\u003c/h2\u003e\n\u003cp\u003eIf you have a function that can take a filename and return a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array,\nsuch as \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eskimage.io.imread\u003c/span\u003e\u003c/code\u003e,\nyou can create a “lazy” version of that function by calling \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.delayed\u003c/span\u003e\u003c/code\u003e on the function itself.\nThis \u003cem\u003enew\u003c/em\u003e function, when handed a filename,\nwill not actually read the file until explicitly asked with the \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003ecompute()\u003c/span\u003e\u003c/code\u003e method:\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003eskimage.io\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003edelayed\u003c/span\u003e\n\n\u003cspan class=\"n\"\u003elazy_imread\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003edelayed\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003ereader\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003elazy_imread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s1\"\u003e'/path/to/file.tif'\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# doesn't actually read the file\u003c/span\u003e\n\u003cspan class=\"n\"\u003earray\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003ereader\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003ecompute\u003c/span\u003e\u003cspan class=\"p\"\u003e()\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# *now* it reads.\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003e(If you have an unusual image format,\nbut you \u003cem\u003edo\u003c/em\u003e have a python function that returns a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array,\nsimply substitute it for \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eskimage.io.imread\u003c/span\u003e\u003c/code\u003e in the example above).\u003c/p\u003e\n\u003cp\u003eWe can create a \u003ca class=\"reference external\" href=\"https://docs.dask.org/en/latest/array.html\" target=\"_blank\" rel=\"noreferrer\"\u003eDask array\u003c/a\u003e of delayed file-readers\nfor \u003cem\u003eall\u003c/em\u003e of the files in our multidimensional experiment using the \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array.from_delayed\u003c/span\u003e\u003c/code\u003e function\nand a \u003ca class=\"reference external\" href=\"https://docs.python.org/3/library/glob.html\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eglob\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e filename pattern\n(\u003cem\u003ethis example assumes that all files are of the same \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eshape\u003c/span\u003e\u003c/code\u003e and \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edtype\u003c/span\u003e\u003c/code\u003e!\u003c/em\u003e):\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003eskimage.io\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003eskimage.io.collection\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003ealphanumeric_key\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003edelayed\u003c/span\u003e\n\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003edask.array\u003c/span\u003e \u003cspan class=\"k\"\u003eas\u003c/span\u003e \u003cspan class=\"nn\"\u003eda\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003eglob\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eglob\u003c/span\u003e\n\n\u003cspan class=\"n\"\u003efilenames\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"nb\"\u003esorted\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003eglob\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"/path/to/experiment/*.tif\"\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e \u003cspan class=\"n\"\u003ekey\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"n\"\u003ealphanumeric_key\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# read the first file to get the shape and dtype\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# ASSUMES THAT ALL FILES SHARE THE SAME SHAPE/TYPE\u003c/span\u003e\n\u003cspan class=\"n\"\u003esample\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003efilenames\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e])\u003c/span\u003e\n\n\u003cspan class=\"n\"\u003elazy_imread\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003edelayed\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# lazy reader\u003c/span\u003e\n\u003cspan class=\"n\"\u003elazy_arrays\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"n\"\u003elazy_imread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003efn\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e \u003cspan class=\"k\"\u003efor\u003c/span\u003e \u003cspan class=\"n\"\u003efn\u003c/span\u003e \u003cspan class=\"ow\"\u003ein\u003c/span\u003e \u003cspan class=\"n\"\u003efilenames\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\n\u003cspan class=\"n\"\u003edask_arrays\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"p\"\u003e[\u003c/span\u003e\n    \u003cspan class=\"n\"\u003eda\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003efrom_delayed\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003edelayed_reader\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003eshape\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"n\"\u003esample\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eshape\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003edtype\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"n\"\u003esample\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003edtype\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n    \u003cspan class=\"k\"\u003efor\u003c/span\u003e \u003cspan class=\"n\"\u003edelayed_reader\u003c/span\u003e \u003cspan class=\"ow\"\u003ein\u003c/span\u003e \u003cspan class=\"n\"\u003elazy_arrays\u003c/span\u003e\n\u003cspan class=\"p\"\u003e]\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# Stack into one large dask.array\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eda\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003edask_arrays\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003eaxis\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eshape\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# (nfiles, nz, ny, nx)\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# in jupyter notebook the repr of a dask stack provides a useful visual:\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003e\u003cimg alt=\"image: dask array html representation\" src=\"../../_images/dask_repr.png\"\u003e\u003c/p\u003e\n\u003cp\u003e\u003cem\u003eNo data has been read from disk yet!\u003c/em\u003e\u003c/p\u003e\n\u003cp\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e is capable of consuming Dask arrays,\nso you can simply call \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari.view_image\u003c/span\u003e\u003c/code\u003e on this \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003estack\u003c/span\u003e\u003c/code\u003e and behind the scenes,\nDask will take care of reading the data from disk\nand handing a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array to \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e each time a new timepoint or channel is requested.\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003enapari\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# specify contrast_limits and is_pyramid=False with big data\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# to avoid unnecessary computations\u003c/span\u003e\n\u003cspan class=\"n\"\u003enapari\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eview_image\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003econtrast_limits\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003cspan class=\"mi\"\u003e2000\u003c/span\u003e\u003cspan class=\"p\"\u003e],\u003c/span\u003e \u003cspan class=\"n\"\u003emultiscale\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"kc\"\u003eFalse\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003e\u003cem\u003eNote: providing the\u003c/em\u003e \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003econtrast_limits\u003c/span\u003e\u003c/code\u003e \u003cem\u003eand\u003c/em\u003e \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003emultiscale\u003c/span\u003e\u003c/code\u003e \u003cem\u003earguments prevents\u003c/em\u003e \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e \u003cem\u003efrom trying to calculate the data min/max, which can take an extremely long time with big data.\nSee \u003ca class=\"reference external\" href=\"https://github.com/napari/napari/issues/736\" target=\"_blank\" rel=\"noreferrer\"\u003enapari issue #736\u003c/a\u003e for further discussion.\u003c/em\u003e\u003c/p\u003e\n\u003c/div\u003e\n\u003cdiv class=\"section\" id=\"make-your-life-easier-with-dask-image\"\u003e\n\u003ch2\u003emake your life easier with \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask-image\u003c/span\u003e\u003c/code\u003e\u003ca class=\"headerlink\" href=\"#make-your-life-easier-with-dask-image\" title=\"Permalink to this headline\"\u003e¶\u003c/a\u003e\u003c/h2\u003e\n\u003cp\u003eThis pattern for creating a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array\u003c/span\u003e\u003c/code\u003e from image data\nhas been previously described in an \u003ca class=\"reference external\" href=\"https://blog.dask.org/2019/06/20/load-image-data\" target=\"_blank\" rel=\"noreferrer\"\u003eexcellent blog post\u003c/a\u003e by John Kirkham.\nIt is a common-enough pattern that John created a useful library (\u003ca class=\"reference external\" href=\"https://github.com/dask/dask-image\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask-image\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e)\nthat does all this for you,\nprovided your image format can be read by the \u003ca class=\"reference external\" href=\"https://github.com/soft-matter/pims\" target=\"_blank\" rel=\"noreferrer\"\u003e\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003epims\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e (Python Image Sequence) reader\n(if not, see note above about providing your own reader function with \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.delayed\u003c/span\u003e\u003c/code\u003e).\u003c/p\u003e\n\u003cp\u003eUsing \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask-image\u003c/span\u003e\u003c/code\u003e, \u003cem\u003eall\u003c/em\u003e of the above code can be simplified to 5 lines:\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003enapari\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask_image.imread\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"/path/to/experiment/*.tif\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003enapari\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eview_image\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003econtrast_limits\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\u003cspan class=\"mi\"\u003e2000\u003c/span\u003e\u003cspan class=\"p\"\u003e],\u003c/span\u003e \u003cspan class=\"n\"\u003eis_pyramid\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"kc\"\u003eFalse\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003e\u003cimg alt=\"image: mCherry-H2B showing chromosome separation during mitosis. Collected on a lattice light sheet microscope\" src=\"../../_images/dask1.gif\"\u003e\u003c/p\u003e\n\u003cdiv class=\"section\" id=\"side-note-regarding-higher-dimensional-datasets\"\u003e\n\u003ch3\u003e\u003cstrong\u003eside note regarding higher-dimensional datasets\u003c/strong\u003e\u003ca class=\"headerlink\" href=\"#side-note-regarding-higher-dimensional-datasets\" title=\"Permalink to this headline\"\u003e¶\u003c/a\u003e\u003c/h3\u003e\n\u003cp\u003eIn the above example, it would be quite common to have a 5+ dimensional dataset\n(e.g. different timepoints \u003cem\u003eand\u003c/em\u003e channels represented among the 3D TIFF files in a folder).\nA standard approach to deal with that sort of thing in \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e would be to \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003ereshape\u003c/span\u003e\u003c/code\u003e the array after instantiation.\nWith Dask, reshaping arrays can \u003cem\u003esometimes\u003c/em\u003e lead to unexpected read events if you’re not careful.\u003c/p\u003e\n\u003cp\u003eFor example:\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask_image.imread\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s1\"\u003e'/path/to/experiment/*.tif'\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eshape\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# -\u0026gt; something like (1200, 64, 256, 280)\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003ecompute\u003c/span\u003e\u003cspan class=\"p\"\u003e()\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# incurs a single file read\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# if there were two channels in that folder you might try to do this:\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003ereshape\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e2\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e600\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e64\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e256\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e280\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# but now trying to read just the first timepoint in the first channel:\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003ecompute\u003c/span\u003e\u003cspan class=\"p\"\u003e()\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# incurs 600 read events!\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003eWe will update this post as best-practices emerge,\nbut one possible solution to this is to avoid reshaping\nDask arrays altogether, by constructing multiple Dask arrays using \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask-image\u003c/span\u003e\u003c/code\u003e, and then using\n\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003eda.stack\u003c/span\u003e\u003c/code\u003e to combine them:\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask_image.imread\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003edask.array\u003c/span\u003e \u003cspan class=\"k\"\u003eas\u003c/span\u003e \u003cspan class=\"nn\"\u003eda\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# # instead of this\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# stack = imread('/path/to/experiment/*.tif')\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# stack = stack.reshape(2, 600, 64, 256, 280)\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# stack[0, 0].compute()  # incurs 600 read events!\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# do something like this:\u003c/span\u003e\n\u003cspan class=\"n\"\u003efile_pattern\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"s2\"\u003e\"/path/to/experiment/*ch\u003c/span\u003e\u003cspan class=\"si\"\u003e{}\u003c/span\u003e\u003cspan class=\"s2\"\u003e*.tif\"\u003c/span\u003e\n\u003cspan class=\"n\"\u003echannels\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003efile_pattern\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eformat\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003ei\u003c/span\u003e\u003cspan class=\"p\"\u003e))\u003c/span\u003e \u003cspan class=\"k\"\u003efor\u003c/span\u003e \u003cspan class=\"n\"\u003ei\u003c/span\u003e \u003cspan class=\"ow\"\u003ein\u003c/span\u003e \u003cspan class=\"nb\"\u003erange\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003enchannels\u003c/span\u003e\u003cspan class=\"p\"\u003e)]\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eda\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003echannels\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eshape\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# (2, 600, 64, 256, 280)\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003ecompute\u003c/span\u003e\u003cspan class=\"p\"\u003e()\u003c/span\u003e  \u003cspan class=\"c1\"\u003e# incurs a single file read\u003c/span\u003e\n\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003c/div\u003e\n\u003c/div\u003e\n\u003cdiv class=\"section\" id=\"processing-data-with-dask-array-map-blocks\"\u003e\n\u003ch2\u003eprocessing data with \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array.map_blocks\u003c/span\u003e\u003c/code\u003e\u003ca class=\"headerlink\" href=\"#processing-data-with-dask-array-map-blocks\" title=\"Permalink to this headline\"\u003e¶\u003c/a\u003e\u003c/h2\u003e\n\u003cp\u003eAs previously mentioned,\nsometimes it is desirable to process data prior to viewing.\nWe’ll take as an example\na series of TIFF files acquired on a lattice-light-sheet microscope.\nA typical workflow might be to deskew, deconvolve, and perhaps crop\nor apply some channel registration prior to viewing.\u003c/p\u003e\n\u003cp\u003eWith \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array.map_blocks\u003c/span\u003e\u003c/code\u003e we can apply any function that accepts a \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enumpy\u003c/span\u003e\u003c/code\u003e array\nand returns a modified array to all the images in our \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array\u003c/span\u003e\u003c/code\u003e.\nIt will be evaluated lazily, when requested (in this case, by \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e);\nwe do not have to wait for it to process the entire dataset.\u003c/p\u003e\n\u003cp\u003eHere is an example of a script that will take a folder of raw tiff files,\nand lazily read, deskew, deconvolve, crop,\nand display them, \u003cem\u003eon demand\u003c/em\u003e as you move the \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e dimensions sliders around.\u003c/p\u003e\n\u003cdiv class=\"highlight-python notranslate\"\u003e\u003cdiv class=\"highlight\"\u003e\u003cpre\u003e\u003cspan\u003e\u003c/span\u003e\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003enapari\u003c/span\u003e\n\u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"nn\"\u003epycudadecon\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003efunctools\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003epartial\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003eskimage\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eio\u003c/span\u003e\n\u003cspan class=\"kn\"\u003efrom\u003c/span\u003e \u003cspan class=\"nn\"\u003edask_image.imread\u003c/span\u003e \u003cspan class=\"kn\"\u003eimport\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# load stacks with dask_image, and psf with skimage\u003c/span\u003e\n\u003cspan class=\"n\"\u003estack\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"/path/to/experiment/*.tif\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003epsf\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003eio\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eimread\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"/path/to/psf.tif\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# prepare some functions that accept a numpy array\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# and return a processed array\u003c/span\u003e\n\n\u003cspan class=\"k\"\u003edef\u003c/span\u003e \u003cspan class=\"nf\"\u003elast3dims\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003ef\u003c/span\u003e\u003cspan class=\"p\"\u003e):\u003c/span\u003e\n    \u003cspan class=\"c1\"\u003e# this is just a wrapper because the pycudadecon function\u003c/span\u003e\n    \u003cspan class=\"c1\"\u003e# expects ndims==3 but our blocks will have ndim==4\u003c/span\u003e\n    \u003cspan class=\"k\"\u003edef\u003c/span\u003e \u003cspan class=\"nf\"\u003efunc\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003earray\u003c/span\u003e\u003cspan class=\"p\"\u003e):\u003c/span\u003e\n        \u003cspan class=\"k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"n\"\u003ef\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003earray\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e0\u003c/span\u003e\u003cspan class=\"p\"\u003e])[\u003c/span\u003e\u003cspan class=\"kc\"\u003eNone\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"o\"\u003e...\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\n    \u003cspan class=\"k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"n\"\u003efunc\u003c/span\u003e\n\n\n\u003cspan class=\"k\"\u003edef\u003c/span\u003e \u003cspan class=\"nf\"\u003ecrop\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003earray\u003c/span\u003e\u003cspan class=\"p\"\u003e):\u003c/span\u003e\n    \u003cspan class=\"c1\"\u003e# simple cropping function\u003c/span\u003e\n    \u003cspan class=\"k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"n\"\u003earray\u003c/span\u003e\u003cspan class=\"p\"\u003e[:,\u003c/span\u003e \u003cspan class=\"mi\"\u003e2\u003c/span\u003e\u003cspan class=\"p\"\u003e:,\u003c/span\u003e \u003cspan class=\"mi\"\u003e10\u003c/span\u003e\u003cspan class=\"p\"\u003e:\u003c/span\u003e\u003cspan class=\"o\"\u003e-\u003c/span\u003e\u003cspan class=\"mi\"\u003e20\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"p\"\u003e:\u003c/span\u003e\u003cspan class=\"mi\"\u003e500\u003c/span\u003e\u003cspan class=\"p\"\u003e]\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# https://docs.python.org/3.8/library/functools.html#functools.partial\u003c/span\u003e\n\u003cspan class=\"n\"\u003edeskew\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003elast3dims\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003epartial\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003epycudadecon\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003edeskew_gpu\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003eangle\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"mf\"\u003e31.5\u003c/span\u003e\u003cspan class=\"p\"\u003e))\u003c/span\u003e\n\u003cspan class=\"n\"\u003edeconv\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003elast3dims\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003epartial\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003epycudadecon\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003edecon\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003epsf\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"n\"\u003epsf\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003ebackground\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"mi\"\u003e10\u003c/span\u003e\u003cspan class=\"p\"\u003e))\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# note: this is done in two steps just as an example...\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# in reality pycudadecon.decon also has a deskew argument\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# map and chain those functions across all dask blocks\u003c/span\u003e\n\u003cspan class=\"n\"\u003edeskewed\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003estack\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003emap_blocks\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003edeskew\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003edtype\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"uint16\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003edeconvolved\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003edeskewed\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003emap_blocks\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003edeconv\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003edtype\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"float32\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\u003cspan class=\"n\"\u003ecropped\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003edeconvolved\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003emap_blocks\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"n\"\u003ecrop\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"n\"\u003edtype\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"s2\"\u003e\"float32\"\u003c/span\u003e\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\n\u003cspan class=\"c1\"\u003e# put the resulting dask array into napari.\u003c/span\u003e\n\u003cspan class=\"c1\"\u003e# (don't forget the contrast limits and is_pyramid==False !)\u003c/span\u003e\n\u003cspan class=\"n\"\u003ev\u003c/span\u003e \u003cspan class=\"o\"\u003e=\u003c/span\u003e \u003cspan class=\"n\"\u003enapari\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003eview_image\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\n    \u003cspan class=\"n\"\u003ecropped\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\n    \u003cspan class=\"n\"\u003econtrast_limits\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"p\"\u003e[\u003c/span\u003e\u003cspan class=\"mi\"\u003e90\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e1500\u003c/span\u003e\u003cspan class=\"p\"\u003e],\u003c/span\u003e\n    \u003cspan class=\"n\"\u003eis_pyramid\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"kc\"\u003eFalse\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\n    \u003cspan class=\"n\"\u003endisplay\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"mi\"\u003e3\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e\n    \u003cspan class=\"n\"\u003escale\u003c/span\u003e\u003cspan class=\"o\"\u003e=\u003c/span\u003e\u003cspan class=\"p\"\u003e(\u003c/span\u003e\u003cspan class=\"mi\"\u003e3\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e1\u003c/span\u003e\u003cspan class=\"p\"\u003e,\u003c/span\u003e \u003cspan class=\"mi\"\u003e1\u003c/span\u003e\u003cspan class=\"p\"\u003e),\u003c/span\u003e\n\u003cspan class=\"p\"\u003e)\u003c/span\u003e\n\n\u003cspan class=\"n\"\u003enapari\u003c/span\u003e\u003cspan class=\"o\"\u003e.\u003c/span\u003e\u003cspan class=\"n\"\u003erun\u003c/span\u003e\u003cspan class=\"p\"\u003e()\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003c/div\u003e\n\u003cp\u003eOf course, the GUI isn’t as responsive as it would be if you had processed the data up front\nand loaded the results into RAM and viewed them in \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e (it’s doing a lot of work after all!),\nbut it’s surprisingly usable,\nand allows you to preview the result of a relatively complex processing pipeline \u003cem\u003eon-the-fly\u003c/em\u003e,\nfor arbitrary timepoints/channels, while storing \u003cem\u003eonly\u003c/em\u003e the raw data on disk.\u003c/p\u003e\n\u003cp\u003e\u003cimg alt=\"image: same dataset, demonstrating on-the-fly read → deskew → deconvolve → crop\" src=\"../../_images/dask2.gif\"\u003e\u003c/p\u003e\n\u003cp\u003eThis workflow is very much patterned after \u003ca class=\"reference external\" href=\"https://blog.dask.org/2019/08/09/image-itk\" target=\"_blank\" rel=\"noreferrer\"\u003eanother great post by John Kirkham, Matthew Rocklin, and Matthew McCormick\u003c/a\u003e\nthat describes a similar image processing pipeline using \u003ca class=\"reference external\" href=\"https://itk.org/\" target=\"_blank\" rel=\"noreferrer\"\u003eITK\u003c/a\u003e.\n\u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003enapari\u003c/span\u003e\u003c/code\u003e simply sits at the end of this lazy processing chain,\nready to show you the result on demand!\u003c/p\u003e\n\u003c/div\u003e\n\u003cdiv class=\"section\" id=\"further-reading\"\u003e\n\u003ch2\u003eFurther Reading\u003ca class=\"headerlink\" href=\"#further-reading\" title=\"Permalink to this headline\"\u003e¶\u003c/a\u003e\u003c/h2\u003e\n\u003cp\u003e\u003ca class=\"reference external\" href=\"https://docs.dask.org/en/latest/delayed.html\" target=\"_blank\" rel=\"noreferrer\"\u003eDocumentation on dask.delayed\u003c/a\u003e\u003c/p\u003e\n\u003cp\u003e\u003ca class=\"reference external\" href=\"https://blog.dask.org/2019/06/20/load-image-data\" target=\"_blank\" rel=\"noreferrer\"\u003eDask working notes on dask-image\u003c/a\u003e\u003c/p\u003e\n\u003cp\u003e\u003ca class=\"reference external\" href=\"https://blog.dask.org/2019/08/09/image-itk\" target=\"_blank\" rel=\"noreferrer\"\u003eDask working notes on image processing with \u003ccode class=\"docutils literal notranslate\"\u003e\u003cspan class=\"pre\"\u003edask.array.map_blocks\u003c/span\u003e\u003c/code\u003e\u003c/a\u003e\u003c/p\u003e\n\u003c/div\u003e\n\u003c/div\u003e\n\n    \u003cscript type=\"text/x-thebe-config\"\u003e\n    {\n        requestKernel: true,\n        binderOptions: {\n            repo: \"binder-examples/jupyter-stacks-datascience\",\n            ref: \"master\",\n        },\n        codeMirrorConfig: {\n            theme: \"abcdef\",\n            mode: \"python\"\n        },\n        kernelOptions: {\n            kernelName: \"python3\",\n            path: \"./tutorials/applications\"\n        },\n        predefinedOutput: true\n    }\n    \u003c/script\u003e\n    \u003cscript\u003ekernelName = 'python3'\u003c/script\u003e\n      ","pageHeaders":[{"href":"#using-dask-delayed-to-load-images","text":"using dask.delayed to load images"},{"href":"#make-your-life-easier-with-dask-image","text":"make your life easier with dask-image"},{"href":"#processing-data-with-dask-array-map-blocks","text":"processing data with dask.array.map_blocks"},{"href":"#further-reading","text":"Further Reading"}]}},"__N_SSG":true},"page":"/[[...parts]]","query":{"parts":["tutorials","applications","dask.html"]},"buildId":"wooSd9gyc-BMXR2p6Uzhj","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>